setwd("/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain")
library(reticulate)
library(Seurat)
library(scCustomize)
library(tidyverse)
library(plotly)
library(SeuratExtend)
library(readxl)
library(anndata)
library(SCP)
library(sceasy)
library(SeuratWrappers)
library(monocle3)
library(sccomp)
library(speckle)
library(Nebulosa)
library(ggplot2)
library(viridis)
library(dplyr)
library(forcats)
library(RColorBrewer)
library(paletteer)
library(khroma)

set.seed(42)
fb_seurat <- readRDS("fb_seurat_v25.RDS")
micro_sub <- readRDS('micro_sub.RDS')
ExN_sub <- readRDS("ExN_sub.RDS")
neuro_sub <- readRDS('neuro_sub.RDS')
oligo_sub <- readRDS("oligo_sub.RDS")
astro_sub <- readRDS("astro_sub.RDS")

# This seurat object was generated by a combination of Megana Shivakumar, Richard Li and John Cornelius
# The seurat object was created from per-sample seurat objects that were then merged, had SCT run on them followed by scDeed
fb_seurat <- readRDS("/gscratch/kawaldorflab/jcorn427/fet_brain/fet_brain_seurat.rds")

DimPlot_scCustom(fb_seurat)

# Harmony integration
fb_seurat <- IntegrateLayers(
  object = fb_seurat, method = HarmonyIntegration,
  orig.reduction = "pca", new.reduction = "harmony"
)

fb_seurat <- FindNeighbors(fb_seurat, reduction = "harmony", dims = 1:30)
fb_seurat <- FindClusters(fb_seurat, resolution = 0.45, cluster.name = "harmony_clusters") #using this resolution to keep pre-harmony # of clusters
fb_seurat <- RunUMAP(fb_seurat, reduction = "harmony", dims = 1:30, reduction.name = "umap.harmony")

DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", group.by = "harmony_clusters", label = TRUE)

# Dot plots for highest expressed genes

fb_seurat <- PrepSCTFindMarkers(fb_seurat)

all_markers <- FindAllMarkers(object = fb_seurat) %>%
  Add_Pct_Diff() 

top_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 7, named_vector = FALSE,
                                   make_unique = TRUE, rank_by = "avg_log2FC")
plots <- Clustered_DotPlot(seurat_object = fb_seurat, features = top_markers, flip = TRUE, x_lab_rotate = 90, elbow_kmax = 30)
plots[[1]]
plots <- Clustered_DotPlot(seurat_object = fb_seurat, features = top_markers, flip = TRUE, x_lab_rotate = 90, k = 28)

top_20_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 20, data_frame = TRUE,
                                      rank_by = "avg_log2FC")
write.csv(top_20_markers, file = "top_20_markers.csv")


# top_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 10, named_vector = FALSE,
#                                    make_unique = TRUE, rank_by = "pct_diff")
# plots <- Clustered_DotPlot(seurat_object = fb_seurat, features = top_markers, flip = TRUE, x_lab_rotate = 90, elbow_kmax = 30)
# plots[[1]]

saveRDS(fb_seurat, file = "fb_seurat_harmony.RDS")
fb_seurat <- readRDS("fb_seurat_harmony.RDS")

# Convert to anndata for scVelo analysis using scCustomize

as.anndata(fb_seurat, file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain", file_name = 'fb_adata.h5ad', assay = "SCT") # Specify the assay you want to convert


# Generating dot plots, remember to switch to RNA assay
DefaultAssay(object = fb_seurat) <- "RNA"
# Remake with new labels
DotPlot_scCustom(fb_seurat, features = c('PTPRC', 'CD14', 'ITGAM', 'CD68', 'CD163', 'CD80', 'CD86', 'FCGR2A', 'MRC1', 'P2RY12', 'HEXB', 'SALL1', "AIF1", 'IL1B'), x_lab_rotate = TRUE)

DotPlot_scCustom(fb_seurat, features = c('AGT', 'OGT', 'FAM107A'), x_lab_rotate = TRUE)

DotPlot_scCustom(fb_seurat, features = c('KCNJ10', 'KCNJ16', 'ATP1A2'), x_lab_rotate = TRUE)

DotPlot_scCustom(fb_seurat, features = c("VIM", "PTPRC", "S100B"), x_lab_rotate = TRUE)

DotPlot_scCustom(fb_seurat, features = c("MPZ", "MAG", "MBP", "NGF", "BDNF"), x_lab_rotate = TRUE)

DotPlot_scCustom(fb_seurat, features = c("NES", "GFAP", "VIM", "EGFR", 'MKI67', 'ASPM', 'TOP2A', 'CDK1', 'OLIG2', 'PDGFRA', 'CSPG4', 'SOX10', 'ENSMMUG00000056728', 'PCDH15', "CD9"), x_lab_rotate = TRUE)

DotPlot_scCustom(fb_seurat, features = c("SOX6", 'BMP4', 'GPR17', 'TNS3', 'FYN', 'PCDH15', "NEU4"), x_lab_rotate = TRUE)

DotPlot_scCustom(fb_seurat, features = c("TCF7L2", 'ITPR2', 'CEMIP2', 'CASR', 'GPR17'), x_lab_rotate = TRUE)

DotPlot_scCustom(fb_seurat, features = c("MAL", 'MOG', 'PLP1', 'OPALIN', 'SERINC5'), x_lab_rotate = TRUE)

DotPlot_scCustom(fb_seurat, features = c("FAR1", 'PMP22', 'MBP', 'KLK6'), x_lab_rotate = TRUE)

DotPlot_scCustom(fb_seurat, features = c("GRM3", 'JPH4', 'APOE'), x_lab_rotate = TRUE)

DotPlot_scCustom(fb_seurat, features = c("SST", 'PVALB', 'CHAT'), x_lab_rotate = TRUE)

DotPlot_scCustom(fb_seurat, features = c("CALB1"), x_lab_rotate = TRUE)

DotPlot_scCustom(fb_seurat, features = c('MRC1', 'CD163', 'ARG1'), x_lab_rotate = TRUE)
# CD206 is MRC1

DotPlot_scCustom(fb_seurat, features = c('GFAP', 'AQP4', 'ENTREP1', 'ENSMMUG00000059505', 'ENSMMUG00000064922', 'ALDOC', 'ACBD7', 'ADIRF', 'PAX6', 'DKK3', 'SLCO1C1', 'NPNT'), x_lab_rotate = TRUE)

DotPlot_scCustom(fb_seurat, features = c('ZBBX', 'ADGB', 'DNAI3', 'NKX6-1', 'PAX6', 'DKK3', 'WNT4', 'BHMT2', 'FOXA1', 'ARX', 'UCMA', 'VTN'), x_lab_rotate = TRUE)
# BHMT is BHMT2 in Macaque

DotPlot_scCustom(fb_seurat, features = c('PROX1', 'GRIN2B', 'GABRG3', 'HTR4', 'PVALB', 'SST', 'VIP'), x_lab_rotate = TRUE)

DotPlot_scCustom(fb_seurat, features = c('CD3D', 'CD8A', 'CD4', 'CD3G', 'ITK', 'S100A4', 'ICOS'), x_lab_rotate = TRUE)

VlnPlot(fb_seurat, features = c("GABRG3", "SATB2", "GRIN2B", "GRIN2A", "SLC17A6", "SLC17A8", "FEZF2"), stack = TRUE, flip = TRUE)
VlnPlot(fb_seurat, features = c("TCF7L1", "SOX3", "SALL1"), stack = TRUE, flip = TRUE)
VlnPlot_scCustom(fb_seurat, features = "TCF7L1")


DefaultAssay(object = fb_seurat) <- "SCT"
FeaturePlot_scCustom(fb_seurat, features = c("SATB2", "PVALB", "SST", "VIP"), reduction = "umap.harmony")

Plot_Density_Joint_Only(fb_seurat, features = c('RELN', 'TBR1', 'GABRG3', 'GABRA1'), reduction = "umap.harmony")
# HAR1F not annotated in Macaque

Plot_Density_Joint_Only(fb_seurat, features = c('PAX6', 'HES1', 'SLC1A3', 'CRYAB', 'NR4A1'), reduction = "umap.harmony")
# SOX2 is not annotated in Macaque

p1 <- Plot_Density_Joint_Only(fb_seurat, features = c('TBR1', 'BCL11B'), reduction = "umap.harmony")
# CTIP2 is BCL11B

p2 <- Plot_Density_Custom(fb_seurat, features = 'SATB2', reduction = "umap.harmony")

p1 + p2

VlnPlot_scCustom(fb_seurat, features = "EOMES")
DefaultAssay(object = fb_seurat) <- "SCT"
FeaturePlot_scCustom(fb_seurat, features = "DDX3Y", reduction = "umap.harmony")

clust_12_markers <- FindMarkers(fb_seurat, ident.1 = "12")


#### Subcluster microglia clusters ####
levels(fb_seurat)
Idents(fb_seurat) <- "harmony_clusters"
clustersToSubset <- c("0", "17")
data.sub <- subset(fb_seurat, idents = clustersToSubset)

DimPlot_scCustom(data.sub, reduction = "umap.harmony")


#Run sctransform
data.sub <- SCTransform(object=data.sub, vst.flavor="v2",  conserve.memory=T, return.only.var.genes=F)

# Run PCA and UMAP
DefaultAssay(data.sub) <- "SCT"
data.sub <- RunPCA(data.sub)
ElbowPlot(data.sub, ndims = 50)
data.sub <- RunUMAP(data.sub, dims = 1:20)

data.sub <- FindNeighbors(data.sub, dims = 1:20)
data.sub <- FindClusters(data.sub, resolution = 0.25)

DimPlot_scCustom(data.sub)

DotPlot_scCustom(data.sub, features = c("GFAP", "ALDH1L1", "S100B", "AQP4", "ITGAM", "TMEM119", "P2RY12", "AIF1", "PTPRC", "BIN1", "CSF1R", "CX3CR1", "CD68", "CD80", "SALL1", "HAVCR2", "CD86", "SLCO2B1", "MRC1", "CD163"), x_lab_rotate = TRUE)




levels(data.sub)
Idents(data.sub) <- "seurat_clusters"

# renaming microglia and astrocytes subclusters
new.cluster.ids <- c("CD80+MG", "CD80+MG", "CD80-MG", "AIF1+MG", "Astro", "CD80-MG", "Astro")
# new.cluster.ids <- c("CD80+MG", "CD80-MG", "AIF1+MG", "Astro", "CD80-MG", "Astro")
names(new.cluster.ids) <- levels(data.sub)
data.sub <- RenameIdents(data.sub, new.cluster.ids)
data.sub$merged <- Idents(data.sub)
DimPlot_scCustom(data.sub)

saveRDS(data.sub, file = "astro_micro_sub.RDS")


# Add microglia and astrocyte cell labels back into main UMAP
CellsMetaTrim <- subset(data.sub@meta.data, select = c("merged"))
fb_seurat <- AddMetaData(fb_seurat, CellsMetaTrim)

meta.data <- fb_seurat@meta.data

meta.data <- meta.data %>% mutate(clusters = coalesce(merged, harmony_clusters))
clusters <- subset(meta.data, select = c("clusters"))
fb_seurat <- AddMetaData(fb_seurat, clusters)

fb_seurat$merged <- NULL


head(fb_seurat@meta.data)
levels(fb_seurat)
Idents(fb_seurat) <- "clusters"

DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", repel = TRUE, label.box = TRUE)

saveRDS(fb_seurat, file = "fb_seurat_v2.RDS")


# 3D UMAP
fb_3D <- fb_seurat
fb_3D <- RunUMAP(fb_3D, reduction = "harmony", dims = 1:20, n.components = 3L, reduction.name = "umap.harmony_3D")

head(Embeddings(object = fb_3D, reduction = "umap.harmony_3D"))

plot.data <- FetchData(object = fb_3D, vars = c("umapharmony3D_1", "umapharmony3D_2", "umapharmony3D_3", "clusters"))

plot.data$label <- paste(rownames(plot.data))

fig <- plot_ly(data = plot.data, 
               x = ~umapharmony3D_1, y = ~umapharmony3D_2, z = ~umapharmony3D_3, 
               color = ~clusters, 
               colors = c("lightseagreen",
                          "gray50",
                          "darkgreen",
                          "red4",
                          "red",
                          "turquoise4",
                          "black",
                          "yellow4",
                          "royalblue1",
                          "lightcyan3",
                          "peachpuff3",
                          "khaki3",
                          "gray20",
                          "orange2",
                          "royalblue4",
                          "yellow3",
                          "gray80",
                          "darkorchid1",
                          "lawngreen",
                          "plum2",
                          "darkmagenta"),
               type = "scatter3d", 
               mode = "markers", 
               marker = list(size = 5, width=2), # controls size of points
               text=~clusters, #This is that extra column we made earlier for which we will use for cell ID
               hoverinfo="text") #When you visualize your plotly object, hovering your mouse pointer over a point shows cell names
fig
head(plot.data)


# renaming fb clusters
levels(fb_seurat)
new.cluster.ids <- c("CD80+MG", "CD80-MG", "AIF1+MG" ,"Astro"  , "iN1"    ,   "OPC2"   ,    "OPC1" ,      "OPC0"   ,    "iN2" ,     
                     "Astro1"    ,   "Astro2"  ,     "8"   ,    "TBR1+N"   ,    "Astro3"  ,    "SATB2+N" ,     "12" ,     "13",      "OL1",     
                     "EP"  ,    "16"  ,    "BCAS1+OL"  ,    "19"  ,    "IPC"  ,    "COP"  ,    "OL2"   ,   "CD3+T" ,     "EB")



# new.cluster.ids <- c("CD80+MG",  "AIF1+MG",  "Astro",   "CD80-MG",   "iN1" ,     "OPC2"  ,   "OPC1"  ,   "OPC0" ,    "iN2"  ,    "Astro1"  ,      "Astro2",       
#                      "8" ,       "TBR1+N",   "Astro3" ,      "SATB2+N",  "12",       "13"  ,     "OL1" ,     "EP"   ,    "16"  ,     "BCAS1+OL", "19",      
#                      "IPC" ,     "COP"  ,    "OL2" ,     "CD3+T" ,   "EB")
names(new.cluster.ids) <- levels(fb_seurat)
fb_seurat <- RenameIdents(fb_seurat, new.cluster.ids)
DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

fb_seurat$cell_type_v1 <- Idents(fb_seurat)


# renaming fb clusters
levels(fb_seurat)
new.cluster.ids <- c("M1-MG" , "M2-MG",  "aMG",  "Astro" ,   "iN1"  ,    "OPC2"  ,   "OPC1" ,    "OPC0",    
                     "iN2" ,     "Astro1",   "Astro2" ,  "8" ,       "TBR1+N",   "Astro3",   "SATB2+N",  "12" ,     
                     "13" ,      "OL1"  ,    "EP"  ,     "16"  ,     "BCAS1+OL" ,"19"  ,     "IPC" ,     "COP" ,    
                     "OL2" ,     "CD3+T" ,   "EB")

# new.cluster.ids <- c("M1-MG",  "AIF1+MG",  "Astro",  "M2-MG",    "iN1",      "OPC2",     "OPC1",     "OPC0",     "iN2",      "Astro1",   "Astro2",  
#                      "8",        "TBR1+N",   "Astro3",   "SATB2+N",  "12",       "13",       "OL1",      "EP",       "16",       "BCAS1+OL", "19",      
#                      "IPC" ,     "COP",      "OL2",      "CD3+T" ,   "EB")
names(new.cluster.ids) <- levels(fb_seurat)
fb_seurat <- RenameIdents(fb_seurat, new.cluster.ids)
DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

fb_seurat$cell_type_v2 <- Idents(fb_seurat)

clustersToSubset <- c("M1-MG", "M2-MG", 'Astro', 'aMG')
data.sub <- subset(fb_seurat, idents = clustersToSubset)
data.sub <- SCTransform(object=data.sub, vst.flavor="v2",  conserve.memory=T, return.only.var.genes=F)
data.sub <- RunPCA(data.sub)
ElbowPlot(data.sub, ndims = 50)
data.sub <- RunUMAP(data.sub, dims = 1:20)


DimPlot_scCustom(data.sub, reduction = "umap", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)
DimPlot_scCustom(micro_sub, reduction = "umap", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

DotPlot_scCustom(data.sub, features = c("GFAP", "ALDH1L1", "S100B", "AQP4", "ITGAM", "TMEM119", "P2RY12", "AIF1", "PTPRC", "BIN1", "CSF1R", "CX3CR1", "CD68", "CD80", "SALL1", "HAVCR2", "CD86", "SLCO2B1", "MRC1", "CD163"), x_lab_rotate = TRUE)

DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)
fb_seurat$cell_type_v3 <- Idents(fb_seurat)
saveRDS(fb_seurat, file = "fb_seurat_v3.RDS")
DotPlot_scCustom(fb_seurat, features = c("DCX", "STMN1", "GABRA1"), x_lab_rotate = TRUE)


#### Subcluster iN clusters ####
levels(fb_seurat)
Idents(fb_seurat) <- "cell_type_v3"
clustersToSubset <- c("iN1", "iN2")
data.sub <- subset(fb_seurat, idents = clustersToSubset)

DimPlot_scCustom(data.sub, reduction = "umap.harmony")

# Yes, actually, we're doing this
#Run sctransform
data.sub <- SCTransform(object=data.sub, vst.flavor="v2",  conserve.memory=T, return.only.var.genes=F)

# Run PCA and UMAP
DefaultAssay(data.sub) <- "SCT"
data.sub <- RunPCA(data.sub)
ElbowPlot(data.sub, ndims = 50)
data.sub <- RunUMAP(data.sub, dims = 1:20)

data.sub <- FindNeighbors(data.sub, dims = 1:20)
data.sub <- FindClusters(data.sub, resolution = 0.15)

DimPlot_scCustom(data.sub)

# Dot plots for highest expressed genes
DefaultAssay(object = data.sub) <- "SCT"
data.sub <- PrepSCTFindMarkers(data.sub)

all_markers <- FindAllMarkers(object = data.sub) %>%
  Add_Pct_Diff() 

top_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 7, named_vector = FALSE,
                                   make_unique = TRUE, rank_by = "avg_log2FC")
plots <- Clustered_DotPlot(seurat_object = data.sub, features = top_markers, flip = TRUE, x_lab_rotate = 90, elbow_kmax = 30)
plots[[1]]
plots <- Clustered_DotPlot(seurat_object = data.sub, features = top_markers, flip = TRUE, x_lab_rotate = 90, k = 5)
DefaultAssay(object = data.sub) <- "RNA"
DotPlot_scCustom(data.sub, features = c("ENSMMUG00000015103"), x_lab_rotate = TRUE)
DotPlot_scCustom(data.sub, features = c("FCER1A", 'IL4', 'KIT', "CPA3",'MS4A2', 'PTPRC'), x_lab_rotate = TRUE)
DotPlot_scCustom(data.sub, features = c('CD34', 'MAMU-DRA', 'CD38', 'CD19', 'CD33', 'TFRC', 'PTPRC'), x_lab_rotate = TRUE)

DotPlot_scCustom(data.sub, features = c("SST"), x_lab_rotate = TRUE)

top_20_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 20, data_frame = TRUE,
                                      rank_by = "avg_log2FC")
write_csv(top_20_markers, file = "top_20_iN_sub.csv")

levels(data.sub)

# renaming sub clusters
levels(data.sub)
new.cluster.ids <- c("iN_0", "iN_1", "iN_2", "MGE-IN", "iN_4")
names(new.cluster.ids) <- levels(data.sub)
data.sub <- RenameIdents(data.sub, new.cluster.ids)
DimPlot_scCustom(data.sub, reduction = "umap", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

data.sub$merged <- Idents(data.sub)




DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

# Add labels back into main UMAP
CellsMetaTrim <- subset(data.sub@meta.data, select = c("merged"))
fb_seurat_2 <- AddMetaData(fb_seurat, CellsMetaTrim)

meta.data <- fb_seurat_2@meta.data

meta.data <- meta.data %>% mutate(clusters = coalesce(merged, cell_type_v2))
clusters_2 <- subset(meta.data, select = c("clusters"))
fb_seurat_2 <- AddMetaData(fb_seurat_2, clusters_2)

fb_seurat_2$merged <- NULL


head(fb_seurat_2@meta.data)
levels(fb_seurat_2)
Idents(fb_seurat_2) <- "clusters"

DimPlot_scCustom(fb_seurat_2, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

saveRDS(fb_seurat_2, 'fb_seurat_v4.RDS')

# Pull top markers for each cluster of the main UMAP to facilitate cluster calling
DefaultAssay(object = fb_seurat) <- "SCT"
fb_seurat <- PrepSCTFindMarkers(fb_seurat)

all_markers <- FindAllMarkers(object = fb_seurat) %>%
  Add_Pct_Diff() 

top_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 7, named_vector = FALSE,
                                   make_unique = TRUE, rank_by = "avg_log2FC")
plots <- Clustered_DotPlot(seurat_object = fb_seurat, features = top_markers, flip = TRUE, x_lab_rotate = 90, elbow_kmax = 30)
plots[[1]]
plots <- Clustered_DotPlot(seurat_object = fb_seurat, features = top_markers, flip = TRUE, x_lab_rotate = 90, k = 25)

top_30_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 30, data_frame = TRUE,
                                      rank_by = "avg_log2FC")
write_csv(top_30_markers, file = "top_30_genes_table.csv")

# Extract count matrix from seurat object

count_matrix <- fb_seurat[["RNA"]]$counts
count_matrix <- as.data.frame(count_matrix)

# Work with output from brain scan
identical(rownames(fb_seurat@meta.data), new_adata_for_MapMyCells_10xWholeHumanBrain_CCN202210140_HierarchicalMapping_UTC_1756914202087$cell_id[1:nrow(fb_seurat@meta.data)])

fb_seurat <- AddMetaData(fb_seurat, new_adata_for_MapMyCells_10xWholeHumanBrain_CCN202210140_HierarchicalMapping_UTC_1756914202087[,c(3,6,9)])
p1 <- DimPlot_scCustom(fb_seurat, reduction = "umap.harmony")

p2 <- DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", group.by = "supercluster_name", label = TRUE, repel = TRUE)
DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", group.by = "supercluster_name", label = TRUE, repel = TRUE)

options(ggrepel.max.overlaps = Inf)

options(ggrepel.max.overlaps = Inf)

p1 +p2

p3 <- DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", group.by = "cluster_name")

p3
p1
p2




#### Rename clusters to catch up with Kristina cluster calls ####
Idents(fb_seurat) <- "clusters"
levels(fb_seurat)

initial_clust <- c("ChC0"  ,   "LGE-IN"  ,   "CGE-IN"  ,   "MGE-IN" ,  "Merge?"  ,   "M1-MG"  ,  "M2-MG" ,   "aMG" ,    
                   "Astro"  ,  "OPC2"  ,   "OPC1"  ,   "OPC0"  ,   "Astro1"  , "Astro2" ,  "PC"   ,     "VGLUT0" , 
                   "B1 NSC" ,  "VGLUT1",  "VGLUT2"   ,    "EC"  ,     "OL1" ,    "EP" ,      "ChC1"   ,    "BCAS1+OL",
                   "VIP-CRH" ,      "IPC"  ,    "COP"  ,    "OL2" ,     "CD3+T" ,   "EB")

names(initial_clust) <- levels(fb_seurat)
fb_seurat <- RenameIdents(fb_seurat, initial_clust)
DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

fb_seurat$cell_type_v4 <- Idents(fb_seurat)

saveRDS(fb_seurat, file = "fb_seurat_v5.RDS")


#### Subcluster Neurons with VGLUT ####
levels(fb_seurat)
Idents(fb_seurat) <- "cell_type_v4"
clustersToSubset <- c("VGLUT0", "VGLUT1", "ChC1", "VIP-CRH", "ChC0", "CGE-IN", "Merge?", "LGE-IN", "MGE-IN", "IPC", "VGLUT2")
data.sub <- subset(fb_seurat, idents = clustersToSubset)

DimPlot_scCustom(data.sub, reduction = "umap.harmony")

# Yes, actually, we're doing this
#Run sctransform
data.sub <- SCTransform(object=data.sub, vst.flavor="v2",  conserve.memory=T, return.only.var.genes=F)

# Run PCA and UMAP
DefaultAssay(data.sub) <- "SCT"
data.sub <- RunPCA(data.sub)
ElbowPlot(data.sub, ndims = 50)
data.sub <- RunUMAP(data.sub, dims = 1:30)

data.sub <- FindNeighbors(data.sub, dims = 1:30)
data.sub <- FindClusters(data.sub, resolution = 0.25)



DimPlot_scCustom(data.sub)
DimPlot_scCustom(data.sub, group.by = "cell_type_v4")


# Dot plots for highest expressed genes
DefaultAssay(object = neuron.sub) <- "SCT"
data.sub <- PrepSCTFindMarkers(data.sub)

all_markers <- FindAllMarkers(object = neuron.sub) %>%
  Add_Pct_Diff() 

top_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 7, named_vector = FALSE,
                                   make_unique = TRUE, rank_by = "avg_log2FC")
plots <- Clustered_DotPlot(seurat_object = data.sub, features = top_markers, flip = TRUE, x_lab_rotate = 90, k = 14)
plots[[1]]

top_all_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = Inf, data_frame = TRUE,
                                      rank_by = "avg_log2FC")

top_all_markers <- top_all_markers[!grepl("^ENSMMUG", rownames(top_all_markers)), ]
write.csv(top_all_markers, file = "neuron_all_markers_ranked.csv")


FeaturePlot_scCustom(data.sub, features = c('TBR1', 'SLC17A6', 'GRIN2A', 'DCX' ,'GRIN2B'))
FeaturePlot_scCustom(neuron.sub, features = c('GRIN2B'))
FeaturePlot_scCustom(neuron.sub, features = c('VIP', 'CRH', 'CHRNA2'))
FeaturePlot_scCustom(data.sub, features = c('LHX6', 'NXPH1', 'NXPH2', 'KCNC2', 'SCGN', 'DLGAP2', 'GRIP2', 'RPH3A'))
FeaturePlot_scCustom(data.sub, features = c('CHODL', 'SCGN', 'GRIP1', 'CACNG5', 'DCX'), num_columns = 3)
FeaturePlot_scCustom(data.sub, features = c('LHX6', 'NKX2-1'))
FeaturePlot_scCustom(data.sub, features = c('TAC3', 'KISS1'))
FeaturePlot_scCustom(data.sub, features = c('MKI67', 'EOMES')) # EOMES is TBR2
FeaturePlot_scCustom(data.sub, features = c('SLC17A7', 'CAMK2A', 'CALB1', 'NEUROD6', 'HPCAL4', 'HCN2'))
FeaturePlot_scCustom(neuron.sub, features = c("EOMES", "TBR1"))
FeaturePlot_scCustom(neuron.sub, features = c("DCX", "TBR1"))
FeaturePlot_scCustom(neuron.sub, features = c("SLC17A7", "SLC17A6", "SLC17A8", "GRIN1", "GRIA2", "GRIA3", "CAMK2A", "CAMK2B"), num_columns = 4)
FeaturePlot_scCustom(neuron.sub, features = c("GAD1", "GAD2", "SLC32A1", "GFAP", "OLIG2", "AQP4"))
FeaturePlot(neuron.sub, features = c("GAD1", "GAD2", "GRIN2B"), blend = TRUE)


saveRDS(data.sub, file = "neuron_sub.RDS")

# Convert to anndata for scVelo analysis using scCustomize

# scCustomize::as.anndata(data.sub, file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain", file_name = 'neuron_adata.h5ad', assay = "SCT") # Specify the assay you want to convert

# Convert to anndata for scVelo analysis using SeuratExtend
Seu2Adata(data.sub)
# Save the AnnData object to a local file
neuron_adata_path <- file.path("/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain", "neuron_adata.h5ad")
adata.Save(neuron_adata_path)

# renaming sub clusters
levels(neuron.sub)
new.cluster.ids <- c("0" , "1" , "2" , "3" ,"4" , "5",  "6",  "7",  "8" , "MKI67 EB" , "10" ,"11", "12")
names(new.cluster.ids) <- levels(neuron.sub)
data.sub <- RenameIdents(neuron.sub, new.cluster.ids)
DimPlot_scCustom(neuron.sub, reduction = "umap", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

data.sub$merged <- Idents(data.sub)




DimPlot_scCustom(fb_seurat_2, reduction = "umap.harmony", group.by = "clusters_2", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

# Add labels back into main UMAP
CellsMetaTrim <- subset(data.sub@meta.data, select = c("merged"))
fb_seurat_2 <- AddMetaData(fb_seurat, CellsMetaTrim)

meta.data <- fb_seurat_2@meta.data

rows_to_replace <- grepl("^[0-9]+$", meta.data$merged)
meta.data$merged[rows_to_replace] <- NA


meta.data <- meta.data %>% mutate(clusters = coalesce(merged, cell_type_v4))
clusters_2 <- subset(meta.data, select = c("clusters"))
fb_seurat_2 <- AddMetaData(fb_seurat_2, clusters_2)

fb_seurat_2$merged <- NULL


head(fb_seurat_2@meta.data)




#### Subcluster OPCs, OLs and COP ####
levels(fb_seurat)
Idents(fb_seurat) <- "cell_type_v9"
clustersToSubset <- c("OPC1", "OPC2", "OPC0", "BCAS1+OL", "COP", "OL1", "OL2")
data.sub <- subset(fb_seurat, idents = clustersToSubset)

DimPlot_scCustom(data.sub, reduction = "umap.harmony")

# Yes, actually, we're doing this
#Run sctransform
data.sub <- SCTransform(object=data.sub, vst.flavor="v2",  conserve.memory=T, return.only.var.genes=F)

# Run PCA and UMAP
DefaultAssay(data.sub) <- "SCT"
data.sub <- RunPCA(data.sub)
ElbowPlot(data.sub, ndims = 50)
data.sub <- RunUMAP(data.sub, dims = 1:30)

data.sub <- FindNeighbors(data.sub, dims = 1:30)
data.sub <- FindClusters(data.sub, resolution = 0.25)



p1 <- DimPlot_scCustom(data.sub)
p2 <- DimPlot_scCustom(data.sub, group.by = "cell_type_v9")
p1+p2

# Dot plots for highest expressed genes
DefaultAssay(object = data.sub) <- "SCT"
data.sub <- PrepSCTFindMarkers(data.sub)

all_markers <- FindAllMarkers(object = data.sub) %>%
  Add_Pct_Diff() 

top_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 7, named_vector = FALSE,
                                   make_unique = TRUE, rank_by = "avg_log2FC")
plots <- Clustered_DotPlot(seurat_object = data.sub, features = top_markers, flip = TRUE, x_lab_rotate = 90, k = 12)
plots[[1]]

top_20_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 20, data_frame = TRUE,
                                      rank_by = "avg_log2FC")
write.csv(top_20_markers, file = "OPC_OL_COP_top_20_markers.csv")
saveRDS(data.sub, file = "OPC_OL_COP_sub.RDS")

# Convert to anndata for scVelo analysis using SeuratExtend
Seu2Adata(data.sub)
# Save the AnnData object to a local file
OPC_OL_COP_adata_path <- file.path("/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain", "OPC_OL_COP_adata.h5ad")
adata.Save(OPC_OL_COP_adata_path)


#### Subcluster Astros ####
levels(fb_seurat)
Idents(fb_seurat) <- "cell_type_v4"
clustersToSubset <- c("B1 NSC", "Astro1", "Astro2")
data.sub <- subset(fb_seurat, idents = clustersToSubset)

DimPlot_scCustom(data.sub, reduction = "umap.harmony")

# Yes, actually, we're doing this
#Run sctransform
data.sub <- SCTransform(object=data.sub, vst.flavor="v2",  conserve.memory=T, return.only.var.genes=F)

# Run PCA and UMAP
DefaultAssay(data.sub) <- "SCT"
data.sub <- RunPCA(data.sub)
ElbowPlot(data.sub, ndims = 50)
data.sub <- RunUMAP(data.sub, dims = 1:20)

data.sub <- FindNeighbors(data.sub, dims = 1:20)
data.sub <- FindClusters(data.sub, resolution = 0.25)



p1 <- DimPlot_scCustom(astro.sub)
p2 <- DimPlot_scCustom(astro.sub, group.by = "cell_type_v4")
p1+p2

# Dot plots for highest expressed genes
DefaultAssay(object = astro.sub) <- "SCT"
astro.sub <- PrepSCTFindMarkers(astro.sub)

all_markers <- FindAllMarkers(object = astro.sub) %>%
  Add_Pct_Diff() 

top_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 7, named_vector = FALSE,
                                   make_unique = TRUE, rank_by = "avg_log2FC")
plots <- Clustered_DotPlot(seurat_object = data.sub, features = top_markers, flip = TRUE, x_lab_rotate = 90, k = 7)
plots[[1]]

all_markers_ranked <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = Inf, data_frame = TRUE,
                                      rank_by = "avg_log2FC")
all_markers_ranked <- all_markers_ranked[!grepl("^ENSMMUG", rownames(all_markers_ranked)), ]

write.csv(all_markers_ranked, file = "astro_all_markers_ranked.csv")

saveRDS(data.sub, file = "astro_sub.RDS")

# Convert to anndata for scVelo analysis using SeuratExtend
Seu2Adata(data.sub)
# Save the AnnData object to a local file
astro_adata_path <- file.path("/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain", "astro_adata.h5ad")
adata.Save(astro_adata_path)

#### Re-do micro sub ####
DefaultAssay(fb_seurat) <- "SCT"
clustersToSubset <- c("M2-MG", "Astro", "aMG", "M1-MG")
data.sub <- subset(fb_seurat, idents = clustersToSubset)

DimPlot_scCustom(data.sub, reduction = "umap.harmony")

# Yes, actually, we're doing this
#Run sctransform
data.sub <- SCTransform(object=data.sub, vst.flavor="v2",  conserve.memory=T, return.only.var.genes=F)

# Run PCA and UMAP
DefaultAssay(data.sub) <- "SCT"
data.sub <- RunPCA(data.sub)
ElbowPlot(data.sub, ndims = 50)
data.sub <- RunUMAP(data.sub, dims = 1:20)

data.sub <- FindNeighbors(data.sub, dims = 1:20)
data.sub <- FindClusters(data.sub, resolution = 0.1)



p1 <- DimPlot_scCustom(data.sub)
p2 <- DimPlot_scCustom(data.sub, group.by = "cell_type_v10")
p1+p2

# Dot plots for highest expressed genes
DefaultAssay(object = data.sub) <- "SCT"
data.sub <- PrepSCTFindMarkers(data.sub)

all_markers <- FindAllMarkers(object = data.sub) %>%
  Add_Pct_Diff() 

top_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 7, named_vector = FALSE,
                                   make_unique = TRUE, rank_by = "avg_log2FC")
plots <- Clustered_DotPlot(seurat_object = data.sub, features = top_markers, flip = TRUE, x_lab_rotate = 90, k = 4)
plots[[1]]

top_20_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 20, data_frame = TRUE,
                                      rank_by = "avg_log2FC")
write.csv(top_20_markers, file = "micro_top_20_markers.csv")


# Convert to anndata for scVelo analysis using SeuratExtend
Seu2Adata(data.sub)
# Save the AnnData object to a local file
micro_adata_path <- file.path("/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain", "micro_adata.h5ad")
adata.Save(micro_adata_path)
saveRDS(data.sub, file = "micro_sub.RDS")




#### Work with output from Allen Brain Atlas ####
# Read in xlsx stuff
micro_allen <- read_excel("micro_hierarchical_mapping_detailed_with_names.xlsx")
oligo_allen <- read_excel("2025 Sep 13_OPC-OLIGO_hierarchical_mapping_with_names.xlsx")
astro_allen <- read_excel("astro_hierarchical_mapping_detailed_with_names.xlsx")
neuron_allen <- read_excel("neuron_hierarchical_mapping_detailed_with_names.xlsx")

# Get metadata from main object
meta.data <- fb_seurat@meta.data
meta.data <- rownames_to_column(meta.data, var = "Barcode")

# Generate neuron dataframe to merge in
neuron_allen$Barcode <- neuron_allen$cell_id
neuron_col <- neuron_allen[, c("SUPC.name", "Barcode"), drop = FALSE]

# Add metadata column and coalesce
meta_data_2 <- left_join(x = meta.data, y = neuron_col, by = "Barcode")
meta_data_2$neuron <- coalesce(meta_data_2$SUPC.name, meta_data_2$cell_type_v4)

# Generate micro dataframe to merge in
micro_allen$Barcode <- micro_allen$cell_id
micro_col <- micro_allen[, c("SUPC.name", "Barcode"), drop = FALSE]

# Add metadata column and coalesce
meta_data_2 <- left_join(x = meta_data_2, y = micro_col, by = "Barcode")
meta_data_2$micro <- coalesce(meta_data_2$SUPC.name.y, meta_data_2$neuron)

# Generate oligo dataframe to merge in
oligo_allen$Barcode <- oligo_allen$cell_id
oligo_col <- oligo_allen[, c("SUPC.name", "Barcode"), drop = FALSE]

# Add metadata column and coalesce
meta_data_2 <- left_join(x = meta_data_2, y = oligo_col, by = "Barcode")
meta_data_2$oligo <- coalesce(meta_data_2$SUPC.name, meta_data_2$micro)

# Generate astro dataframe to merge in
astro_allen$Barcode <- astro_allen$cell_id
astro_col <- astro_allen[, c("SUPC.name", "Barcode"), drop = FALSE]

# Add metadata column and coalesce
meta_data_2 <- left_join(x = meta_data_2, y = astro_col, by = "Barcode")
meta_data_2$allen_annotations <- coalesce(meta_data_2$SUPC.name.y.y, meta_data_2$oligo)

# Add back into main UMAP and plot
meta_data_2<- meta_data_2 %>% select(Barcode, allen_annotations)

fb_seurat <- AddMetaData(fb_seurat, meta_data_2)

p1 <- DimPlot_scCustom(fb_seurat, reduction = "umap.harmony")

p2 <- DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", group.by = "allen_annotations", label = TRUE, repel = TRUE)
p1+p2
DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", group.by = "allen_annotations", label = TRUE, repel = TRUE)

options(ggrepel.max.overlaps = Inf)


#### Repeat for all the subsets ####
# Micro_sub

micro_sub <- AddMetaData(micro_sub, micro_col)
DimPlot_scCustom(micro_sub, reduction = "umap", group.by = "SUPC.name", label = TRUE, repel = TRUE)

# Oligo_sub
oligo.sub <- AddMetaData(oligo.sub, oligo_col)
DimPlot_scCustom(oligo.sub, reduction = "umap", group.by = "SUPC.name", label = TRUE, repel = TRUE)

# Astro_sub
astro.sub <- AddMetaData(astro.sub, astro_col)
DimPlot_scCustom(astro.sub, reduction = "umap", group.by = "SUPC.name", label = TRUE, repel = TRUE)

# Neuron_sub
neuron.sub <- AddMetaData(neuron.sub, neuron_col)
DimPlot_scCustom(neuron.sub, reduction = "umap", group.by = "SUPC.name", label = TRUE, repel = TRUE)

#### DDX3Y Microglia ####

levels(micro_sub)
unique(micro_sub$clusters)
FeaturePlot_scCustom(micro_sub, features = "DDX3Y", label = TRUE)

#### Random Plots ####

DimPlot_scCustom(fb_seurat, group.by = "cell_type_v4", reduction = "umap.harmony")


DimPlot_scCustom(neuron.sub, reduction = "umap.harmony")


Stacked_VlnPlot(fb_seurat, features = "MKI67", x_lab_rotate = TRUE)

Stacked_VlnPlot(neuron.sub, features = "MKI67", x_lab_rotate = TRUE)

Stacked_VlnPlot(fb_seurat, features = "MKI67", x_lab_rotate = TRUE)

FeaturePlot_scCustom(fb_seurat, features = "MKI67", label = TRUE)

FeaturePlot_scCustom(astro.sub, features = c("SLCO4A1", "VIM", 'NES', "HES1", "GLI3", "TNC", "PRRX1",
                                             "HOPX", "CRYAB", "NR4A1", "SOX2", "PAX6"))


#### Subset cluster 3 of Neuron Sub ####
neuron.sub2 <- FindSubCluster(object = neuron.sub, cluster = "3",graph.name = "SCT_snn", resolution = 0.25, subcluster.name = "sub_3")

DimPlot_scCustom(neuron.sub2, group.by = "sub_3", label = TRUE)

DotPlot_scCustom(
  neuron.sub2,
  features = c("HBA", "HBE1"),
  group.by = "sub_3",
  assay = "RNA",
  scale = FALSE)

FeaturePlot(neuron.sub2, features = c("GRIN2B", "DCX"), label = TRUE, blend = TRUE)

FeaturePlot_scCustom(fb_seurat, features = "ARPP21", label = TRUE, reduction = "umap.harmony")

fb_seurat <- NormalizeData(fb_seurat, assay = "RNA")

# 3_0 is IPC
# Merge 5 and 7, called MGE-IN
# Merge 1,0 CGE-IN1
# 6 is CGE-IN2

#### Get co-expression of GRIN2B 

Idents(neuron.sub2) <- "sub_3"

# Assuming your Seurat object is named 'seurat_object'
# Define your genes of interest
gene1 <- "GRIN2B"
gene2 <- "GAD1"

# Define the minimum expression threshold (e.g., > 0)
coexpression_threshold <- 0.4

# Get all unique cluster identities
all_clusters <- unique(Idents(neuron.sub2))

# Create an empty data frame to store results
coexpression_results <- data.frame(
  Cluster = character(),
  Coexpression_Percentage = numeric(),
  stringsAsFactors = FALSE
)

# Loop through each cluster
for (cluster_id in all_clusters) {
  # Subset the object to only include cells from the current cluster
  cluster_cells <- subset(neuron.sub2, idents = cluster_id)
  
  # Find cells co-expressing both genes within this cluster, above the threshold
  coexpressing_cells_in_cluster <- WhichCells(
    cluster_cells,
    expression = get(gene1) > coexpression_threshold & get(gene2) > coexpression_threshold
  )
  
  # Calculate the percentage within the current cluster
  num_coexpressing <- length(coexpressing_cells_in_cluster)
  total_cells_in_cluster <- ncol(cluster_cells)
  
  percentage_in_cluster <- (num_coexpressing / total_cells_in_cluster) * 100
  
  # Store the results in the data frame
  coexpression_results <- rbind(
    coexpression_results,
    data.frame(
      Cluster = cluster_id,
      Coexpression_Percentage = percentage_in_cluster
    )
  )
}

# Print the results
print(coexpression_results)



FeaturePlot_scCustom(neuron.sub2, features = c("NKX2-1", "LHX6", "SOX6", "NR2F1", "NR2F2", "PROX1", "SP8", "DLX2"), num_columns = 4)
FeaturePlot_scCustom(neuron.sub2, features = c("SST"))
grep("PVALB", Features(neuron.sub2), value = TRUE)


levels(neuron.sub2)

DotPlot_scCustom(neuron.sub2, features = c("NKX2-1", "LHX6", "SOX6", "NR2F1", "NR2F2", "PROX1", "SP8", "DLX1"), flip_axes = TRUE)
DotPlot_scCustom(fb_seurat_2, features = c("NKX2-1", "LHX6", "SOX6", "NR2F1", "NR2F2", "PROX1", "SP8", "DLX1", "GRIN2B", "GAD1", "GAD2", "DLX2", "DLX5", "DLX6", "SST", "NPY", "NOS1"), flip_axes = TRUE, x_lab_rotate = TRUE)


DimPlot_scCustom(neuron.sub2)


#### Add Neuron Sub back into main ####
# Add labels back into main UMAP
CellsMetaTrim <- subset(neuron.sub2@meta.data, select = c("sub_3"))
fb_seurat_2 <- AddMetaData(fb_seurat, CellsMetaTrim)

meta.data <- fb_seurat_2@meta.data

meta.data <- meta.data %>% mutate(temp = coalesce(sub_3, cell_type_v4))
clusters_2 <- subset(meta.data, select = c("temp"))
fb_seurat_2 <- AddMetaData(fb_seurat_2, clusters_2)

fb_seurat_2$sub_3 <- NULL


head(fb_seurat_2@meta.data)
levels(fb_seurat_2)
Idents(fb_seurat_2) <- "temp"

DimPlot_scCustom(fb_seurat_2, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

saveRDS(fb_seurat_2, 'fb_seurat_v6.RDS')

DotPlot_scCustom(fb_seurat, group.by = "cell_type_v5", features = c("NKX2-1", "LHX6", "SOX6", "NR2F1", "NR2F2", "PROX1", "SP8", "DLX1", "GRIN2B", "GAD1", "GAD2", "DLX2", "DLX5", "DLX6", "SST", "NPY", "NOS1", "SLC32A1", "VIP", "CRH", "CCK", "HTR3A", "CALB2", "RELN",
                                           "ACKR3",  'CXCR4', 'NRP1', 'NRP2', 'SP9'), flip_axes = TRUE, x_lab_rotate = TRUE)



#### Rename suclusters on neuron sub ####
Idents(neuron.sub2) <- "sub_3"
levels(neuron.sub2)

initial_clust <- c("4" ,  "CGE-IN1" ,  "8" ,  "CGE-IN1"  , "9" ,  "2" ,  "IPC", "MGE-IN" ,  "CGE-IN2" ,  "11" , "ExN", "MGE-IN",   "12",  "10")

names(initial_clust) <- levels(neuron.sub2)
neuron.sub2 <- RenameIdents(neuron.sub2, initial_clust)
DimPlot_scCustom(neuron.sub2, label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

neuron.sub2$neu_cells_v1 <- Idents(neuron.sub2)

saveRDS(neuron.sub2, file = "neuron.sub_v2.RDS")


# Add labels back into main UMAP
CellsMetaTrim <- subset(neuron.sub2@meta.data, select = c("neu_cells_v1"))
fb_seurat_2 <- AddMetaData(fb_seurat, CellsMetaTrim)

meta.data <- fb_seurat_2@meta.data

meta.data <- meta.data %>% mutate(temp = coalesce(neu_cells_v1, cell_type_v4))
clusters_2 <- subset(meta.data, select = c("temp"))
fb_seurat_2 <- AddMetaData(fb_seurat_2, clusters_2)

fb_seurat_2$neu_cells_v1 <- NULL


head(fb_seurat_2@meta.data)
levels(fb_seurat_2)
Idents(fb_seurat_2) <- "temp"

DimPlot_scCustom(fb_seurat_2, reduction = "umap.harmony", label = FALSE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

#### Rename suclusters on neuron sub again ####
Idents(neuron.sub2) <- "neu_cells_v1"
levels(neuron.sub2)

initial_clust <- c("4" ,      "CGE-IN", "8" ,      "9"   ,    "CGE-IN"  ,     "IPC" ,    "MGE-IN" , "CGE-IN" ,"11",      "ExN" ,    "12",     
                   "10")

names(initial_clust) <- levels(neuron.sub2)
neuron.sub2 <- RenameIdents(neuron.sub2, initial_clust)
DimPlot_scCustom(neuron.sub2, label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

neuron.sub2$neu_cells_v2 <- Idents(neuron.sub2)

saveRDS(neuron.sub2, file = "neuron.sub_v3.RDS")

FeaturePlot_scCustom(neuron.sub2, features = c('SOX2', 'PAX6', 'SLC1A3', 'VIM', 'PDGFD', 'HOPX', 'ITGB5', 'GFAP', 'GLAST', 'BLBP', 'GLI3', "HES1", "PROM1", 'TNC', 'PTPRZ1', 'FAM107A', "FABP7", "MKI67"), num_columns = 4)
FeaturePlot_scCustom(neuron.sub2, features = c('FAM107B', 'MOXD1', 'ANXA1', 'SLC1A3', "LIFR"), num_columns = 3)

#### Rename suclusters on neuron sub again again ####
Idents(neuron.sub2) <- "neu_cells_v2"
levels(neuron.sub2)

initial_clust <- c("4"  ,    "CGE-IN", "8"  ,    "vRG"   ,   "IPC" ,   "MGE-IN", "11"  ,   "ExN" ,   "12" ,   
                   "oRG")

names(initial_clust) <- levels(neuron.sub2)
neuron.sub2 <- RenameIdents(neuron.sub2, initial_clust)
DimPlot_scCustom(neuron.sub2, label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

neuron.sub2$neu_cells_v3 <- Idents(neuron.sub2)

saveRDS(neuron.sub2, file = "neuron.sub_v4.RDS")


FeaturePlot_scCustom(neuron.sub2, features = c('TBR1', 'FOXP2', 'FEZF2', 'BCL11B', 'SOX5', 'SATB2', 'CUX1', 'CUX2', 'POU3F2', 'POU3F3', 'RORB'), num_columns = 4)


#### Rename suclusters on neuron sub again again again####
Idents(neuron.sub2) <- "neu_cells_v3"
levels(neuron.sub2)

initial_clust <- c("IPC-UL" ,     "CGE-IN", "8" ,     "vRG" ,   "IPC" ,   "MGE-IN", "11" ,    "IPC-DL" ,   "12",    
                   "oRG")

names(initial_clust) <- levels(neuron.sub2)
neuron.sub2 <- RenameIdents(neuron.sub2, initial_clust)
DimPlot_scCustom(neuron.sub2, label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

neuron.sub2$neu_cells_v4 <- Idents(neuron.sub2)

saveRDS(neuron.sub2, file = "neuron.sub_v5.RDS")

FeaturePlot_scCustom(neuron.sub2, features = c('SLC17A7', 'CAMK2A'))


# Add labels back into main UMAP
CellsMetaTrim <- subset(neuron.sub2@meta.data, select = c("neu_cells_v2"))
fb_seurat_2 <- AddMetaData(fb_seurat, CellsMetaTrim)

meta.data <- fb_seurat_2@meta.data

meta.data <- meta.data %>% mutate(temp = coalesce(neu_cells_v1, cell_type_v4))
clusters_2 <- subset(meta.data, select = c("temp"))
fb_seurat_2 <- AddMetaData(fb_seurat_2, clusters_2)

fb_seurat_2$neu_cells_v1 <- NULL


head(fb_seurat_2@meta.data)
levels(fb_seurat_2)
Idents(fb_seurat_2) <- "temp"

DimPlot_scCustom(fb_seurat_2, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)


#### Rename labels for main umap ####

levels(fb_seurat_2)
fb_seurat_2$cell_type_v5 <- Idents(fb_seurat_2)
saveRDS(fb_seurat_2, file = "fb_seurat_v6.RDS")


head(fb_seurat@meta.data)

# Dot plots for highest expressed genes

fb_seurat <- PrepSCTFindMarkers(fb_seurat)

all_markers <- FindAllMarkers(object = fb_seurat) %>%
  Add_Pct_Diff() 

top_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 7, named_vector = FALSE,
                                   make_unique = TRUE, rank_by = "avg_log2FC")
plots <- Clustered_DotPlot(seurat_object = fb_seurat, features = top_markers, flip = TRUE, x_lab_rotate = 90)
plots[[1]]
plots <- Clustered_DotPlot(seurat_object = fb_seurat, features = top_markers, flip = TRUE, x_lab_rotate = 90, k = 20)

top_20_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 20, data_frame = TRUE,
                                      rank_by = "avg_log2FC")
write.csv(top_20_markers, file = "top_20_markers.csv")

# Rename ExN to DL-ExN

DotPlot_scCustom(fb_seurat, features = c('SATB2', 'CUX1', 'CUX2', 'BRN2', 'RORB'), x_lab_rotate = TRUE)

# Rename 4 to UL-ExN

DimPlot_scCustom(fb_seurat, reduction = "umap.harmony")

initial_clust <- c("UL-ExN"    ,    "CGE-IN1",  "8" ,       "vRG"   ,     "2"    ,    "IPC"    ,  "MGE-IN" ,  "CGE-IN2" , "11"  ,     "DL-ExN",     
                   "12" ,      "oRG" ,      "M1-MG" ,   "M2-MG" ,   "aMG" ,     "Astro"  ,  "OPC2" ,    "OPC1" ,    "OPC0" ,    "Astro1",  
                   "Astro2",   "PC"  ,     "B1 NSC",   "EC" ,      "OL1" ,     "EP"  ,     "BCAS1+OL", "COP",      "OL2" ,     "CD3+T" ,  
                   "EB")

names(initial_clust) <- levels(fb_seurat)
fb_seurat <- RenameIdents(fb_seurat, initial_clust)
DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

fb_seurat$cell_type_v6 <- Idents(fb_seurat)

saveRDS(fb_seurat, file = "fb_seurat_v7.RDS")

#### Subset MGE-IN ####

mge_in <- subset(fb_seurat, idents = "MGE-IN")

DimPlot_scCustom(mge_in, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

mge_in2 <- FindSubCluster(object = mge_in, cluster = "MGE-IN",graph.name = "SCT_snn", resolution = 0.15, subcluster.name = "sub")

DimPlot_scCustom(mge_in2, reduction = "umap.harmony", group.by = "sub", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

Idents(mge_in2) <- "sub"

mge_in_all_markers <- FindAllMarkers(object = mge_in2) %>%
  Add_Pct_Diff()

mge_in_top_markers <- Extract_Top_Markers(marker_dataframe = mge_in_all_markers, num_genes = 10, named_vector = FALSE,
                                   make_unique = TRUE, rank_by = "avg_log2FC")

DotPlot_scCustom(mge_in2, features = mge_in_top_markers, x_lab_rotate = TRUE)

plots <- Clustered_DotPlot(seurat_object = mge_in2, features = mge_in_top_markers, flip = TRUE, x_lab_rotate = 90)
plots[[1]]

FeaturePlot_scCustom(mge_in2, features = c("TAC3", "UXT", "SST"), reduction = "umap.harmony")

FeaturePlot(mge_in2, features = c("UXT", "DDX3Y"), blend = TRUE, reduction = "umap.harmony")

# Add labels back into main UMAP
CellsMetaTrim <- subset(mge_in2@meta.data, select = c("sub"))
fb_seurat_2 <- AddMetaData(fb_seurat, CellsMetaTrim)

meta.data <- fb_seurat_2@meta.data

meta.data <- meta.data %>% mutate(cell_type_v7 = coalesce(sub, cell_type_v6))
clusters_2 <- subset(meta.data, select = c("cell_type_v7"))
fb_seurat_2 <- AddMetaData(fb_seurat_2, clusters_2)

fb_seurat_2$sub <- NULL


head(fb_seurat_2@meta.data)
levels(fb_seurat_2)
Idents(fb_seurat_2) <- "cell_type_v7"

DimPlot_scCustom(fb_seurat_2, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

saveRDS(fb_seurat_2, file = "fb_seurat_v8.RDS")


#### Look at main UMAP again ####

DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

#### Rename cluster 11 ####
levels(fb_seurat)

initial_clust <- c("UL-ExN" ,  "OPC2"   ,  "CGE-IN1",  "OL2",      "8",        "Astro2",   "OPC1",     "B1 NSC",   "M2-MG",    "BCAS1+OL",
                   "COP"  ,    "EC"   ,   "vRG"   ,   "EP"    ,   "Astro1" ,  "OPC0"  ,   "2"   ,     "IPC"   ,   "PC"   ,    "MGE-IN_0",
                   "M1-MG" ,   "OL1" ,     "CGE-IN2" , "aMG"  ,    "DBC"  ,     "DL-ExN" ,  "CD3+T" ,   "Astro"  ,  "EB"  ,     "MGE-IN_1",
                   "12"  ,     "oRG")

names(initial_clust) <- levels(fb_seurat)
fb_seurat <- RenameIdents(fb_seurat, initial_clust)
DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

fb_seurat$cell_type_v8 <- Idents(fb_seurat)

saveRDS(fb_seurat, file = "fb_seurat_v9.RDS")

#### Update Neuron Sub names ####

DimPlot_scCustom(neuron.sub, label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

#### Merge CGE in main UMAP ####
levels(fb_seurat)

initial_clust <- c("UL-ExN" ,  "OPC2"   ,  "CGE-IN",  "OL2",      "8",        "Astro2",   "OPC1",     "B1 NSC",   "M2-MG",    "BCAS1+OL",
                   "COP"  ,    "EC"   ,   "vRG"   ,   "EP"    ,   "Astro1" ,  "OPC0"  ,   "CGE-IN"   ,     "IPC"   ,   "PC"   ,    "MGE-IN_0",
                   "M1-MG" ,   "OL1" ,     "CGE-IN" , "aMG"  ,    "DBC"  ,     "DL-ExN" ,  "CD3+T" ,   "Astro"  ,  "EB"  ,     "MGE-IN_1",
                   "12"  ,     "oRG")

names(initial_clust) <- levels(fb_seurat)
fb_seurat <- RenameIdents(fb_seurat, initial_clust)
DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

fb_seurat$cell_type_v9 <- Idents(fb_seurat)

saveRDS(fb_seurat, file = "fb_seurat_v10.RDS")


#### Subset 8 and 12 ####

data.sub <- subset(fb_seurat, idents = c("8", "12", "OPC0", "BCAS1+OL"))

DimPlot_scCustom(data.sub, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

# Dot plots for highest expressed genes

fb_seurat <- PrepSCTFindMarkers(fb_seurat)

all_markers <- FindAllMarkers(object = fb_seurat) %>%
  Add_Pct_Diff() 

not_all_markers <- all_markers[grepl(paste(c("8", "12", "OPC0", "BCAS1\\+OL", "COP"), collapse = "|"), all_markers$cluster), ]

top_markers <- Extract_Top_Markers(marker_dataframe = not_all_markers, num_genes = 7, named_vector = FALSE,
                                   make_unique = TRUE, rank_by = "avg_log2FC")
DotPlot_scCustom(fb_seurat, top_markers, x_lab_rotate = TRUE, flip_axes = TRUE)


plots <- Clustered_DotPlot(seurat_object = data.sub, features = top_markers, flip = TRUE, x_lab_rotate = 90)
plots[[1]]
plots <- Clustered_DotPlot(seurat_object = fb_seurat, features = top_markers, flip = TRUE, x_lab_rotate = 90, k = 20)

top_20_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 20, data_frame = TRUE,
                                      rank_by = "avg_log2FC")
write.csv(top_20_markers, file = "top_20_markers.csv")

DotPlot_scCustom(fb_seurat, features = c("ENSMMUG00000056728", "GPR17", "SOX8"), x_lab_rotate = TRUE, flip_axes = TRUE)

#### Merge COP and BCAS1+OL in main UMAP ####
levels(fb_seurat)

initial_clust <- c("UL-ExN" ,  "OPC2"  ,   "CGE-IN" ,  "OL2"  ,    "8"   ,     "Astro1" ,  "OPC1" ,    "Astro2" ,  "M2-MG",   
                   "COP", "COP" ,     "EC" ,      "vRG"  ,    "EP" ,      "Astro0" ,  "OPC0" ,    "IPC"  ,    "PC" ,     
                   "MGE-IN_0", "M1-MG",    "OL1",      "aMG"  ,    "DBC" ,     "DL-ExN" ,  "CD3+T" ,   "Astro" ,   "EB" ,     
                   "MGE-IN_1", "12"  ,     "oRG")

names(initial_clust) <- levels(fb_seurat)
fb_seurat <- RenameIdents(fb_seurat, initial_clust)
DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

fb_seurat$cell_type_v10 <- Idents(fb_seurat)

saveRDS(fb_seurat, file = "fb_seurat_v11.RDS")


#### Rename stuff ####

DimPlot_scCustom(fb_seurat, group.by = "cell_type_v9", reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

levels(fb_seurat)

initial_clust <- c("UL-ExN",   "OPC2" ,    "CGE-IN" ,  "OL2"  ,    "8"   ,     "AS1" ,  "OPC1" ,    "AS0" ,  "M2-MG",   
                   "COP",      "EC" ,      "vRG" ,     "EP" ,      "AS2" ,  "OPC0"  ,   "IPC" ,     "PC"  ,     "MGE-IN_0",
                   "M1-MG",    "OL1" ,     "aMG" ,     "DBC" ,     "DL-ExN" ,  "CD3+T" ,   "Astro" ,   "EB"   ,    "MGE-IN_1",
                   "12",       "oRG")

names(initial_clust) <- levels(fb_seurat)
fb_seurat <- RenameIdents(fb_seurat, initial_clust)
DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

fb_seurat$cell_type_v11 <- Idents(fb_seurat)

saveRDS(fb_seurat, file = "fb_seurat_v12.RDS")


# Dot plots for highest expressed genes of interest

fb_seurat <- PrepSCTFindMarkers(fb_seurat)

all_markers <- FindAllMarkers(object = fb_seurat) %>%
  Add_Pct_Diff() 

all_markers2 <- all_markers[!grepl("^ENSMMUG", rownames(all_markers)), ]

not_all_markers <- all_markers2[grepl(paste(c("MGE-IN_0", "MGE-IN_1", "8", "12", "Astro", "aMG", "M1-MG", "M2-MG"), collapse = "|"), all_markers2$cluster), ]

top_markers <- Extract_Top_Markers(marker_dataframe = not_all_markers, num_genes = 7, named_vector = FALSE,
                                   make_unique = TRUE, rank_by = "avg_log2FC")
DotPlot_scCustom(fb_seurat, top_markers, x_lab_rotate = TRUE, flip_axes = TRUE)


plots <- Clustered_DotPlot(seurat_object = data.sub, features = top_markers, flip = TRUE, x_lab_rotate = 90)
plots[[1]]
plots <- Clustered_DotPlot(seurat_object = fb_seurat, features = top_markers, flip = TRUE, x_lab_rotate = 90, k = 20)

top_20_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 20, data_frame = TRUE,
                                      rank_by = "avg_log2FC")
write.csv(top_20_markers, file = "top_20_markers.csv")

#### Subset out 8, 12, OPC0 and AS2 ####

data.sub <- subset(fb_seurat, idents = c("8", "12", "OPC0", "AS2"))

DimPlot_scCustom(data.sub, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

DotPlot_scCustom(data.sub, features = c("PDGFRA", "CSPG4", "NG2", "SOX9", "NFIA", "OLIG2", "NFIB", "EOMES", "VIM", "LAMC1"))

#### Merging 8 and 12, renaming to GPC ####

levels(fb_seurat)

initial_clust <- c("UL-ExN" ,  "OPC2"  ,   "CGE-IN" ,  "OL2"  ,    "GPC"    ,    "AS1"  ,    "OPC1"  ,   "AS0"  ,    "M2-MG" ,  
                   "COP"   ,   "EC"   ,    "vRG" ,     "EP"  ,     "AS2" ,     "OPC0"   ,  "IPC"   ,   "PC"  ,     "MGE-IN_0",
                   "M1-MG"  ,  "OL1"    ,  "aMG"   ,   "DBC"  ,    "DL-ExN",   "CD3+T"   , "Astro" ,   "EB"   ,    "MGE-IN_1",
                   "GPC"   ,    "oRG")

names(initial_clust) <- levels(fb_seurat)
fb_seurat <- RenameIdents(fb_seurat, initial_clust)
DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

fb_seurat$cell_type_v12 <- Idents(fb_seurat)

saveRDS(fb_seurat, file = "fb_seurat_v13.RDS")

#### Subset out MGE-IN_0, MGE-IN_1, CGE-IN, DBC ####

data.sub <- subset(fb_seurat, idents = c("MGE-IN_0", "MGE-IN_1", "CGE-IN", "DBC"))

DimPlot_scCustom(data.sub, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

DotPlot_scCustom(data.sub, features = c('LHX6', 'SOX6', 'ERBB4', 'KCNC1', "KCNC2", 'NPY', 'CHRNA2', 'GRIK3'))

# Yes, actually, we're doing this
#Run sctransform
data.sub <- SCTransform(object=data.sub, vst.flavor="v2",  conserve.memory=T, return.only.var.genes=F)

# Run PCA and UMAP
DefaultAssay(data.sub) <- "SCT"
data.sub <- RunPCA(data.sub)
ElbowPlot(data.sub, ndims = 50)
data.sub <- RunUMAP(data.sub, dims = 1:30)

data.sub <- FindNeighbors(data.sub, dims = 1:30)
data.sub <- FindClusters(data.sub, resolution = 0.15)

p1 <- DimPlot_scCustom(data.sub)
p2 <- DimPlot_scCustom(data.sub, group.by = "cell_type_v11")
p1 + p2

#### Subcluster CGE-IN ####
fb_seurat2 <- FindSubCluster(object = fb_seurat, cluster = "CGE-IN",graph.name = "SCT_snn", resolution = 0.25, subcluster.name = "sub_clust")

DimPlot_scCustom(fb_seurat2, reduction = "umap.harmony", group.by = "sub_clust", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)


#### Subset out MGE-IN_0, MGE-IN_1, CGE-IN, DBC ####
Idents(fb_seurat2) <- "sub_clust"


data.sub <- subset(fb_seurat2, idents = c("MGE-INP", "MGE-P", "iCGE-IN", "CGE-P", "CGE-IN_3", "DBC"))

DimPlot_scCustom(data.sub, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

DotPlot_scCustom(data.sub, features = c("NKX2-1", "DLX1", "DLX2", "DLX5", "DLX6", "LHX6", "SOX6", "GRIN2B", "GAD1", "GAD2", "ERBB4", "KCNA10", "KCN2", "NPY", "CHRNA2", "GRIK3", "TAC3", "SP9", "NR2F1", "NR2F2", "PROX1", "CALB2", "SP8", "CCK", "NRP2",
                                        "CXCR4", "ACKR3", "VIP"), flip_axes = TRUE, x_lab_rotate = TRUE, colors_use = c('blue', 'white', 'red'))

#### Rename MGE-IN_0 to MGE-INP, CGE-IN_2 to CGE-P, CGE-IN_0 and CGE-IN_1 to iCGE-IN, MGE-IN_1 to MGE-P #### 

levels(fb_seurat2)

initial_clust <- c("UL-ExN" ,  "OPC2" ,    "iCGE-IN", "OL2" ,     "GPC" ,     "AS1" ,     "iCGE-IN", "OPC1" ,    "AS0",     
                   "M2-MG",    "COP" ,     "EC"   ,    "vRG" ,     "EP" ,      "AS2" ,     "OPC0"  ,   "CGE-P", "IPC",     
                   "PC" ,      "MGE-INP", "M1-MG",    "OL1" ,     "CGE-IN_3", "aMG" ,     "DBC" ,     "DL-ExN",   "CD3+T",   
                   "Astro" ,   "EB" ,      "MGE-P", "oRG")

names(initial_clust) <- levels(fb_seurat2)
fb_seurat2 <- RenameIdents(fb_seurat2, initial_clust)
DimPlot_scCustom(fb_seurat2, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

fb_seurat2$cell_type_v13 <- Idents(fb_seurat2)

saveRDS(fb_seurat2, file = "fb_seurat_v14.RDS")


DimPlot_scCustom(fb_seurat2, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

#### Rename CGE-IN_3 and DBC to iCGE-VIP/CR-IN ####
levels(fb_seurat2)

initial_clust <- c("UL-ExN",   "OPC2"  ,   "iCGE-IN" , "OL2"  ,    "GPC" ,     "AS1"  ,    "OPC1",     "AS0"  ,    "M2-MG",   
                   "COP",      "EC" ,      "vRG" ,     "EP" ,      "AS2" ,     "OPC0",     "CGE-P",    "IPC",      "PC",      
                   "MGE-INP",  "M1-MG" ,   "OL1" ,     "iCGE-VIP/CR-IN", "aMG",      "iCGE-VIP/CR-IN",      "DL-ExN",   "CD3+T",    "Astro",   
                   "EB" ,      "MGE-P" ,   "oRG")

names(initial_clust) <- levels(fb_seurat2)
fb_seurat2 <- RenameIdents(fb_seurat2, initial_clust)
DimPlot_scCustom(fb_seurat2, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

fb_seurat2$cell_type_v14 <- Idents(fb_seurat2)

saveRDS(fb_seurat2, file = "fb_seurat_v15.RDS")

#### Look at the microglia populations ####

data.sub <- subset(fb_seurat, idents = c("M2-MG", "M1-MG", "Astro", "aMG"))

DimPlot_scCustom(data.sub, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

# Yes, actually, we're doing this
#Run sctransform
data.sub <- SCTransform(object=data.sub, vst.flavor="v2",  conserve.memory=T, return.only.var.genes=F)

# Run PCA and UMAP
DefaultAssay(data.sub) <- "SCT"
data.sub <- RunPCA(data.sub)
ElbowPlot(data.sub, ndims = 50)
data.sub <- RunUMAP(data.sub, dims = 1:30)

data.sub <- FindNeighbors(data.sub, dims = 1:30)
data.sub <- FindClusters(data.sub, resolution = 0.1)

p1 <- DimPlot_scCustom(data.sub)
p2 <- DimPlot_scCustom(data.sub, group.by = "cell_type_v14")
p1 + p2

DotPlot_scCustom(data.sub, group.by = "cell_type_v14", features = c("GFAP", "ALDH1L1", "S100B", "AQP4", "ITGAM", "TMEM119", "P2RY12", "AIF1", "PTPRC", "BIN1", "CSF1R", "CX3CR1", "CD68", "CD80", "SALL1", "HAVCR2", "CD86", "SLCO2B1", "MRC1", "CD163"), x_lab_rotate = TRUE, flip_axes = TRUE, colors_use = c("blue", "white", "red"))

#### Trying to subcluster microglia clusters ####

levels(fb_seurat)

new_clust <- c("UL-ExN"  ,       "OPC2"   ,        "iCGE-IN"  ,      "OL2"  ,          "GPC"     ,       "AS1" ,          
               "OPC1"  ,         "AS0"    ,        "MG"   ,       "COP"    ,        "EC"    ,         "vRG" ,          
               "EP"   ,          "AS2"   ,         "OPC0" ,          "CGE-P"  ,        "IPC" ,           "PC"  ,          
               "MGE-INP"  ,      "MG" ,         "OL1" ,           "iCGE-VIP/CR-IN", "MG",            "DL-ExN",  
               "CD3+T" ,         "MG"   ,       "EB"     ,        "MGE-P"  ,        "oRG")

names(new_clust) <- levels(fb_seurat)
fb_seurat <- RenameIdents(fb_seurat, new_clust)
DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

fb_seurat2 <- FindSubCluster(object = fb_seurat, cluster = "MG",graph.name = "SCT_snn", resolution = 0.25, subcluster.name = "sub")

DimPlot_scCustom(fb_seurat2, group.by = "sub", reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

# Dot plots for highest expressed genes of interest

fb_seurat2 <- PrepSCTFindMarkers(fb_seurat2)

Idents(fb_seurat2) <- "sub"

all_markers <- FindAllMarkers(object = fb_seurat2) %>%
  Add_Pct_Diff() 

all_markers2 <- all_markers[!grepl("^ENSMMUG", rownames(all_markers)), ]

not_all_markers <- all_markers2[grepl(paste(c("MG_0", "MG_1", "MG_2", "MG_3", "MG_4"), collapse = "|"), all_markers2$cluster), ]

top_markers <- Extract_Top_Markers(marker_dataframe = not_all_markers, num_genes = 7, named_vector = FALSE,
                                   make_unique = TRUE, rank_by = "avg_log2FC")
DotPlot_scCustom(fb_seurat2, top_markers, x_lab_rotate = TRUE, flip_axes = TRUE)


plots <- Clustered_DotPlot(seurat_object = data.sub, features = top_markers, flip = TRUE, x_lab_rotate = 90)
plots[[1]]
plots <- Clustered_DotPlot(seurat_object = fb_seurat, features = top_markers, flip = TRUE, x_lab_rotate = 90, k = 20)

top_20_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 20, data_frame = TRUE,
                                      rank_by = "avg_log2FC")
write.csv(top_20_markers, file = "top_20_markers.csv")

Idents(fb_seurat2) <- "sub_clust"


data.sub <- subset(fb_seurat2, idents = c("MG_0", "MG_1", "MG_2", "MG_3", "MG_4"))

DimPlot_scCustom(data.sub, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

DotPlot_scCustom(data.sub, features = c("CD68", "AIF1", "IL1B", "CD80", "CD86", "CX3CR1", "SALL1", "TREM2", "CD163", "MRC1", "IL10", "TGFB1", "GFAP", "AQP4", "S100B"), flip_axes = TRUE, x_lab_rotate = TRUE, colors_use = c('blue', 'white', 'red'))

#### Merge all the MG ####

levels(fb_seurat2)

initial_clust <- c("UL-ExN"    ,     "OPC2"  ,         "iCGE-IN" ,       "OL2"  ,          "GPC",            "AS1" ,           "OPC1" ,         
                   "AS0"  ,          "MG"   ,        "COP"         ,   "MG"   ,        "EC"     ,        "vRG"  ,          "EP"  ,          
                   "AS2"  ,          "MG"   ,        "OPC0" ,          "CGE-P"  ,        "IPC"   ,         "PC"  ,           "MGE-INP"  ,     
                   "OL1"   ,         "iCGE-VIP/CR-IN", "MG" ,          "MG"    ,       "DL-ExN"  ,       "CD3+T"      ,    "EB"   ,         
                   "MGE-P"   ,       "oRG")

names(initial_clust) <- levels(fb_seurat2)
fb_seurat2 <- RenameIdents(fb_seurat2, initial_clust)
DimPlot_scCustom(fb_seurat2, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

fb_seurat2$cell_type_v15 <- Idents(fb_seurat2)

saveRDS(fb_seurat2, file = "fb_seurat_v16.RDS")


FeaturePlot_scCustom(fb_seurat, features = c('CD68', 'CD45', 'MHCII', "M2A", "IL1RA", "FTH1", "FTL", "MAMU-DRB1", "ARG1", "IL1"), reduction = "umap.harmony")

DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", group.by = "cell_type_v14")

#### Subcluster AIF1+ cells from MG cluster ####

Percent_Expressing(fb_seurat, "AIF1", threshold = 3)

FeaturePlot_scCustom(fb_seurat, features = "AIF1", reduction = "umap.harmony", na_cutoff = 3)

# Specify your gene and the expression threshold
gene_of_interest <- "AIF1"
threshold <- 3

Idents(fb_seurat) <- "cell_type_v15"

high_aif1_cells <- WhichCells(
  object = fb_seurat,
  idents = "MG",
  expression = AIF1 > 3
)

# Convert the original cluster identities to a character vector
fb_seurat$cell_type_v16 <- as.character(Idents(fb_seurat))

# Check the head of the new metadata column
head(fb_seurat$cell_type_v16)

# Assign the new cluster identity to the high_aif1 cells
fb_seurat$cell_type_v16[high_aif1_cells] <- "AIF1++ MG"

# Set the new_clusters column as the active identity
Idents(fb_seurat) <- "cell_type_v16"

DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", group.by = "cell_type_v16")

saveRDS(fb_seurat, file = "fb_seurat_v17.RDS")




#### Subcluster Astros ####
levels(fb_seurat)
clustersToSubset <- c("GPC", "AS0", "AS1", "AS2")
data.sub <- subset(fb_seurat, idents = clustersToSubset)

DimPlot_scCustom(data.sub, reduction = "umap.harmony")

# Yes, actually, we're doing this
#Run sctransform
data.sub <- SCTransform(object=data.sub, vst.flavor="v2",  conserve.memory=T, return.only.var.genes=F)

# Run PCA and UMAP
DefaultAssay(data.sub) <- "SCT"
data.sub <- RunPCA(data.sub)
ElbowPlot(data.sub, ndims = 50)
data.sub <- RunUMAP(data.sub, dims = 1:20)

data.sub <- FindNeighbors(data.sub, dims = 1:20)
data.sub <- FindClusters(data.sub, resolution = 0.25)

DimPlot_scCustom(data.sub, group.by = "cell_type_v19")

p1 <- DimPlot_scCustom(data.sub)
p2 <- DimPlot_scCustom(data.sub, group.by = "cell_type_v16")
p1+p2

saveRDS(data.sub, file = "astro_sub_v2.RDS")


#### Swap astro 2 and astro 0, sub-cluster MG ####
levels(fb_seurat)

initial_clust <- c("UL-ExN"  ,       "OPC2"   ,        "iCGE-IN"  ,      "OL2" ,           "GPC" ,           "AS1",            "OPC1",          
                   "AS2"  ,          "MG"    ,         "COP"    ,        "EC"     ,        "vRG"    ,        "EP"   ,          "AS0"  ,         
                   "OPC0"     ,      "CGE-P"     ,     "IPC"      ,      "PC"        ,     "MGE-INP"  ,      "OL1"  ,          "iCGE-VIP/CR-IN",
                   "AIF1++ MG"  ,    "DL-ExN"    ,     "CD3+T"  ,        "EB"  ,           "MGE-P"   ,       "oRG")

names(initial_clust) <- levels(fb_seurat)
fb_seurat <- RenameIdents(fb_seurat, initial_clust)
DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

fb_seurat$cell_type_v17 <- Idents(fb_seurat)
saveRDS(fb_seurat, file = "fb_seurat_v18.RDS")


Idents(fb_seurat_2) <- "cell_type_v17"

clustersToSubset <- c("AIF1++ MG", "MG")
data.sub <- subset(fb_seurat_2, idents = clustersToSubset)

DimPlot_scCustom(data.sub, reduction = "umap.harmony")

# Yes, actually, we're doing this
#Run sctransform
data.sub <- SCTransform(object=data.sub, vst.flavor="v2",  conserve.memory=T, return.only.var.genes=F)

# Run PCA and UMAP
DefaultAssay(data.sub) <- "SCT"
data.sub <- RunPCA(data.sub)
ElbowPlot(data.sub, ndims = 50)
data.sub <- RunUMAP(data.sub, dims = 1:20)

data.sub <- FindNeighbors(data.sub, dims = 1:20)
data.sub <- FindClusters(data.sub, resolution = 0.1)

p1 <- DimPlot_scCustom(data.sub)
p2 <- DimPlot_scCustom(data.sub, group.by = "cell_type_v18")
p1+p2
p1
p2
FeaturePlot_scCustom(data.sub, features = "CD80")
FeaturePlot_scCustom(data.sub, features = "CD86")
FeaturePlot_scCustom(data.sub, features = "AIF1")
FeaturePlot_scCustom(data.sub, features = c("ARG1", "MMR", "IL10", "YM1"))

data.sub$sub <- Idents(data.sub)

saveRDS(data.sub, "micro_sub_v2.RDS")



# Add labels back into main UMAP
CellsMetaTrim <- subset(data.sub@meta.data, select = c("sub"))
fb_seurat_2 <- AddMetaData(fb_seurat, CellsMetaTrim)

meta.data <- fb_seurat_2@meta.data

meta.data <- meta.data %>% mutate(cell_type_v18 = coalesce(sub, cell_type_v17))
clusters_2 <- subset(meta.data, select = c("cell_type_v18"))
fb_seurat_2 <- AddMetaData(fb_seurat_2, clusters_2)

fb_seurat_2$sub <- NULL


head(fb_seurat_2@meta.data)
levels(fb_seurat_2)
Idents(fb_seurat_2) <- "cell_type_v18"

DimPlot_scCustom(fb_seurat_2, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

saveRDS(fb_seurat_2, "fb_seurat_v18.RDS")

# Dot plots for highest expressed genes of interest

fb_seurat_2 <- PrepSCTFindMarkers(fb_seurat_2)

Idents(fb_seurat_2) <- "cell_type_v18"

all_markers <- FindAllMarkers(object = fb_seurat_2) %>%
  Add_Pct_Diff() 

all_markers2 <- all_markers[!grepl("^ENSMMUG", rownames(all_markers)), ]

not_all_markers <- all_markers2[grepl(paste(c("0", "1", "2", "3"), collapse = "|"), all_markers2$cluster), ]

not_all_markers <- all_markers2[all_markers2$cluster %in% c(1, 3), ]

top_markers <- Extract_Top_Markers(marker_dataframe = not_all_markers, num_genes = 7, named_vector = FALSE,
                                   make_unique = TRUE, rank_by = "avg_log2FC")
DotPlot_scCustom(fb_seurat_2, top_markers, x_lab_rotate = TRUE, flip_axes = TRUE)


plots <- Clustered_DotPlot(seurat_object = fb_seurat_2, features = top_markers, flip = TRUE, x_lab_rotate = 90)
plots[[1]]
plots <- Clustered_DotPlot(seurat_object = fb_seurat, features = top_markers, flip = TRUE, x_lab_rotate = 90, k = 20)

top_20_markers <- Extract_Top_Markers(marker_dataframe = not_all_markers, num_genes = 20, data_frame = TRUE,
                                      rank_by = "avg_log2FC")
write.csv(top_20_markers, file = "top_20_markers_1_3.csv")

#### CGE and MGE Neural Progenitor Subset ####
clustersToSubset <- c("CGE-P", "iCGE-IN", "iCGE-VIP/CR-IN", "MGE-INP", "MGE-P", "vRG", "oRG")
data.sub <- subset(fb_seurat, idents = clustersToSubset)

DimPlot_scCustom(data.sub, reduction = "umap.harmony")

# Yes, actually, we're doing this
#Run sctransform
data.sub <- SCTransform(object=data.sub, vst.flavor="v2",  conserve.memory=T, return.only.var.genes=F)

# Run PCA and UMAP
DefaultAssay(data.sub) <- "SCT"
data.sub <- RunPCA(data.sub)
ElbowPlot(data.sub, ndims = 50)
data.sub <- RunUMAP(data.sub, dims = 1:30)

data.sub <- FindNeighbors(data.sub, dims = 1:30)
data.sub <- FindClusters(data.sub, resolution = 0.1)


p1 <- DimPlot_scCustom(data.sub)
p2 <- DimPlot_scCustom(data.sub, group.by = "cell_type_v18")
p1+p2
p1
p2

head(fb_seurat_2@meta.data)
saveRDS(fb_seurat_2, "fb_seurat_v19.RDS")

saveRDS(data.sub, "neuron_sub_v6.RDS")

#### Oligo and Astro Subset ####
clustersToSubset <- c("GPC", "OPC0", "OPC1", "OPC2", "COP", "OL1", "OL2", "AS0", "AS1", "AS2")
data.sub <- subset(fb_seurat_2, idents = clustersToSubset)

DimPlot_scCustom(data.sub, reduction = "umap.harmony")

# Yes, actually, we're doing this
#Run sctransform
data.sub <- SCTransform(object=data.sub, vst.flavor="v2",  conserve.memory=T, return.only.var.genes=F)

# Run PCA and UMAP
DefaultAssay(data.sub) <- "SCT"
data.sub <- RunPCA(data.sub)
ElbowPlot(data.sub, ndims = 50)
data.sub <- RunUMAP(data.sub, dims = 1:30)

data.sub <- FindNeighbors(data.sub, dims = 1:30)
data.sub <- FindClusters(data.sub, resolution = 0.1)


p1 <- DimPlot_scCustom(data.sub)
p2 <- DimPlot_scCustom(data.sub, group.by = "cell_type_v18")
p1+p2
p1
p2

saveRDS(data.sub, "Oligo_Astro_sub.RDS")


# Convert to anndata for scVelo analysis using scCustomize
as.anndata(
  x = fb_seurat_2,
  file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain",
  file_name = "fb_adata_2.h5ad"
)

# Convert micro_sub to anndata for scVelo analysis using scCustomize
as.anndata(
  x = micro_sub,
  file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain",
  file_name = "micro_adata.h5ad"
)

# Convert neuron_sub to anndata for scVelo analysis using scCustomize
as.anndata(
  x = neuron.sub,
  file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain",
  file_name = "neuron_adata.h5ad"
)

# Convert oligo_astro_sub to anndata for scVelo analysis using scCustomize
as.anndata(
  x = oligo_astro.sub,
  file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain",
  file_name = "oligo_astro_adata.h5ad"
)

#### Rename micro clusters and incorporate them into the main UMAP ####
levels(micro_sub)

initial_clust <- c("MG1", "MG2", "Mono", "BAM")

names(initial_clust) <- levels(micro_sub)
micro_sub <- RenameIdents(micro_sub, initial_clust)
DimPlot_scCustom(micro_sub,label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

micro_sub$cell_type_v19 <- Idents(micro_sub)

saveRDS(micro_sub, file = "micro_sub_v3.RDS")


Idents(fb_seurat_2) <- "cell_type_v18"

levels(fb_seurat_2)

initial_clust <- c("MG1"   ,           "MG2"   ,           "Mono"    ,          "BAM" ,             "UL-ExN" ,        "OPC2" ,          "iCGE-IN" ,      
                   "OL2" ,           "GPC"   ,         "AS1"   ,         "OPC1"  ,         "AS2" ,           "COP",            "EC",       
                   "vRG"  ,          "EP" ,            "AS0"   ,         "OPC0" ,         "CGE-P" ,         "IPC" ,           "PC" ,           
                   "MGE-INP" ,       "OL1"  ,          "iCGE-VIP/CR-IN" ,"DL-ExN" ,        "CD3+T"  ,        "EB"   ,          "MGE-P" ,        
                   "oRG")

names(initial_clust) <- levels(fb_seurat_2)
fb_seurat_2 <- RenameIdents(fb_seurat_2, initial_clust)
DimPlot_scCustom(fb_seurat_2, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

fb_seurat_2$cell_type_v19 <- Idents(fb_seurat_2)

saveRDS(fb_seurat_2, file = "fb_seurat_v20.RDS")

nrow(fb_seurat_2)

DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

FeaturePlot_scCustom(neuron.sub, features = c("CCK", "VIP"))

FeaturePlot_scCustom(fb_seurat, features = c("CD68", "CD3D", "CD69", "ITGAE"), reduction = "umap.harmony")


#### Final Tweaks ####

fb_seurat <- RenameIdents(fb_seurat, "CD3+T" = "TRM")

levels(fb_seurat)
fb_seurat$cell_type_v19 <- Idents(fb_seurat)
saveRDS(fb_seurat, file = "fb_seurat_v20.RDS")


fb_seurat2 <- FindSubCluster(object = fb_seurat, cluster = "iCGE-VIP/CR-IN",graph.name = "SCT_snn", resolution = 0.25, subcluster.name = "cell_type_v20")

DimPlot_scCustom(fb_seurat2, group.by = "cell_type_v20", reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

Idents(fb_seurat2) <- "cell_type_v20"
clustersToSubset <- c("iCGE-VIP/CR-IN_1", "iCGE-VIP/CR-IN_0")
data.sub <- subset(fb_seurat2, idents = clustersToSubset)

DimPlot_scCustom(data.sub, reduction = "umap.harmony")

FeaturePlot_scCustom(data.sub, features = c("CALB2", "VIP"), reduction = "umap.harmony", label = FALSE, split.by = "cell_type_v20")


levels(fb_seurat)
clustersToSubset <- c("GPC", "OPC2", "OL2", "OPC1", "COP", "OPC0", "OL1")
oligo.sub <- subset(fb_seurat, idents = clustersToSubset)

DimPlot_scCustom(oligo.sub, reduction = "umap.harmony")

#### Renaming more idents ####
fb_seurat <- RenameIdents(fb_seurat, c("CGE-P" = "CGE0", "iCGE-IN" = "CGE1", "iCGE-VIP/CR-IN" = "CGE2", "MGE-P" = "MGE0", "MGE-INP" = "MGE1"))
levels(fb_seurat)
DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

fb_seurat$cell_type_v19 <- Idents(fb_seurat)
saveRDS(fb_seurat, file = "fb_seurat_v20.RDS")

DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29,
                                                                                                                                                   palette = "varibow", shuffle_pal = FALSE), color_seed = 5)


order <- c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1", "GPC", "AS0", "AS1", "AS2", 'OPC0', 'OPC1', 'OPC2', "COP", 'OL1', 'OL2', "IPC", "DL-ExN", "UL-ExN", "Mono", "MG1", "MG2", "MG3", "TRM", "EB", "EC", "PC", "EP")

fb_seurat$cell_type_v19 <- factor(
  x = fb_seurat$cell_type_v19,
  levels = order
)

Idents(fb_seurat) <- "cell_type_v19"

DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = FALSE), color_seed = 5)
                                                                                                                                                                              
saveRDS(fb_seurat, file = "fb_seurat_v20.RDS")

#### Final final Tweaks ####

fb_seurat <- RenameIdents(fb_seurat, "BAM" = "MG3")

fb_seurat <- RenameIdents(fb_seurat, c("MG2" = "MG1"))


levels(fb_seurat)
fb_seurat$cell_type_v19 <- Idents(fb_seurat)
saveRDS(fb_seurat, file = "fb_seurat_v20.RDS")



#### Generating subsets for UMAPs ####
levels(fb_seurat)
clustersToSubset <- c("GPC", "OPC2", "OL2", "OPC1", "COP", "OPC0", "OL1")
oligo.sub <- subset(fb_seurat, idents = clustersToSubset)
saveRDS(oligo.sub, "oligo_sub.RDS")
DimPlot_scCustom(oligo.sub, reduction = "umap.harmony")
DimPlot_scCustom(oligo.sub, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = TRUE), color_seed = 5)

levels(fb_seurat)
clustersToSubset <- c("GPC", "AS0", "AS1", "AS2")
astro.sub <- subset(fb_seurat, idents = clustersToSubset)
saveRDS(astro.sub, 'astro_sub.RDS')
DimPlot_scCustom(astro.sub, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "polychrome", shuffle_pal = TRUE), color_seed = 5)

levels(fb_seurat)
clustersToSubset <- c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1")
neuro.sub <- subset(fb_seurat, idents = clustersToSubset)
saveRDS(neuro.sub, "neuro_sub.RDS")
DimPlot_scCustom(neuro.sub, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = TRUE), color_seed = 5)

levels(fb_seurat)
clustersToSubset <- c("IPC", "DL-ExN", "UL-ExN")
IPC.sub <- subset(fb_seurat, idents = clustersToSubset)
saveRDS(IPC.sub, "IPC_sub.RDS")
DimPlot_scCustom(IPC.sub, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "polychrome", shuffle_pal = TRUE), color_seed = 5)


#### Look at the microglia populations ####

micro.sub <- subset(fb_seurat, idents = c("Mono", "MG1", "MG2", "MG3"))

DimPlot_scCustom(micro.sub, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

# Yes, actually, we're doing this
#Run sctransform
micro.sub <- SCTransform(object=micro.sub, vst.flavor="v2",  conserve.memory=T, return.only.var.genes=F)

# Run PCA and UMAP
DefaultAssay(micro.sub) <- "SCT"
micro.sub <- RunPCA(micro.sub)
ElbowPlot(micro.sub, ndims = 50)
micro.sub <- RunUMAP(micro.sub, dims = 1:30)

micro.sub <- FindNeighbors(micro.sub, dims = 1:30)
micro.sub <- FindClusters(micro.sub, resolution = 0.1)

p1 <- DimPlot_scCustom(micro.sub)
p2 <- DimPlot_scCustom(micro.sub, group.by = "cell_type_v19")
p1 + p2
p2

DimPlot_scCustom(micro.sub, group.by = "cell_type_v19", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = TRUE), color_seed = 5)
saveRDS(micro.sub, "micro_sub.RDS")









#### Re-exporting for trajectory analysis for analysis ####

# Convert to anndata for scVelo analysis using scCustomize
as.anndata(
  x = fb_seurat,
  file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain",
  file_name = "fb_adata_3.h5ad"
)

# Convert micro_sub to anndata for scVelo analysis using scCustomize
as.anndata(
  x = micro_sub,
  file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain",
  file_name = "micro_adata_2.h5ad"
)

# Convert neuron_sub to anndata for scVelo analysis using scCustomize
as.anndata(
  x = neuro_sub,
  file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain",
  file_name = "neuro_adata.h5ad"
)

# Convert oligo_sub to anndata for scVelo analysis using scCustomize
as.anndata(
  x = oligo_sub,
  file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain",
  file_name = "oligo_adata.h5ad"
)

# Convert astro_sub to anndata for scVelo analysis using scCustomize
as.anndata(
  x = astro_sub,
  file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain",
  file_name = "astro_adata.h5ad"
)

#### Trying stuff with micro ####

levels(fb_seurat)
initial_clust <- c("vRG",    "oRG" ,   "CGE0" ,  "CGE1",   "CGE2" ,  "MGE0" ,  "MGE1" ,  "GPC" ,   "AS0" ,   "AS1" ,   "AS2" ,   "OPC0",  
                   "OPC1",   "OPC2",   "COP",    "OL1",    "OL2",    "IPC",    "DL-ExN", "UL-ExN", "MG",   "MG",    "MG",    "MG",   
                   "TRM",    "EB" ,    "EC" ,    "PC" ,    "EP")

names(initial_clust) <- levels(fb_seurat)
fb_seurat <- RenameIdents(fb_seurat, initial_clust)
fb_seurat2 <- FindSubCluster(object = fb_seurat, cluster = "MG",graph.name = "SCT_snn", resolution = 0.2, subcluster.name = "cell_type_v20")


gradient <- c("blue", "white", "red")
micro_plot <- DotPlot_scCustom(fb_seurat2, features = c( "PTPRC", "CD14", "FCGR3", "CD68", "CD163", "AIF1", "C1QB", "C1QC",  "SALL1",
                                                        "CX3CR1", "ITGAL", "MERTK", "STAB1", "P2RY12", "MRC1", "ITGAM", "CSF1R", "CCR2", "TMEM119", "LYVE1", "GPR34", "CD80", "CD86", "IL1B", "IL6",
                                                        "TREM2", "APOE", "CTSD", "ARG1", "CHI3L1"), 
                               group.by = "cell_type_v20", remove_axis_titles = FALSE,  flip_axes = TRUE, x_lab_rotate = TRUE, colors_use = gradient) + labs(x = "Gene", y = "Cluster")

micro_plot  +
  guides(size = guide_legend(title = "Percent expressed", order = 2),
         color = guide_colorbar(title = "Avg. Expression (scaled)", order = 1)) +
  scale_size_continuous(breaks = c(0, 20, 40, 60),
                        limits = c(0, 100)) +
  scale_color_gradientn(
    colors = gradient,
    breaks = c(-1, 0, 1)
  ) +
  theme(
    legend.title = element_text(size = 14),
    legend.text = element_text(size = 12)
  )

Idents(fb_seurat2) <- "cell_type_v20"

micro.sub <- subset(fb_seurat2, idents = c("MG_0", "MG_1", "MG_2", "MG_3"))

DimPlot_scCustom(micro.sub, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

micro_plot <- DotPlot_scCustom(micro.sub, features = c( "PTPRC", "CD14", "FCGR3", "CD68", "CD163", "AIF1", "C1QB", "C1QC",  "SALL1",
                                                         "CX3CR1", "ITGAL", "MERTK", "STAB1", "P2RY12", "MRC1", "ITGAM", "CSF1R", "CCR2", "TMEM119", "LYVE1", "GPR34", "CD80", "CD86", "IL1B", "IL6",
                                                         "TREM2", "APOE", "CTSD", "ARG1", "CHI3L1"), 
                               group.by = "cell_type_v20", remove_axis_titles = FALSE,  flip_axes = TRUE, x_lab_rotate = TRUE, colors_use = gradient) + labs(x = "Gene", y = "Cluster")

micro_plot  +
  guides(size = guide_legend(title = "Percent expressed", order = 2),
         color = guide_colorbar(title = "Avg. Expression (scaled)", order = 1)) +
  scale_size_continuous(breaks = c(0, 20, 40, 60),
                        limits = c(0, 100)) +
  scale_color_gradientn(
    colors = gradient,
    breaks = c(-1, 0, 1)
  ) +
  theme(
    legend.title = element_text(size = 14),
    legend.text = element_text(size = 12)
  )


#### Merge MG_0, MG_1, MG_2, label them MG. Rename MG_3 to AIF1+MG ####

Idents(micro.sub)
micro.sub <- RenameIdents(micro.sub, c("MG_0" = "MG", "MG_1" = "MG", "MG_2" = "MG", "MG_3" = "AIF1+MG"))
micro.sub$cell_type_v20 <- Idents(micro.sub)

micro_plot <- DotPlot_scCustom(micro.sub, features = c( "CD68", "CD163", "AIF1", "C1QB", "C1QC",  "SALL1",
                                                        "CX3CR1", "P2RY12", "MRC1", "CSF1R", "TMEM119", "LYVE1", "GPR34", "CD80", "CD86", "IL1B", "IL6",
                                                        "TREM2", "APOE", "CTSD"), 
                               group.by = "cell_type_v20", remove_axis_titles = FALSE,  flip_axes = TRUE, x_lab_rotate = TRUE, colors_use = gradient) + labs(x = "Gene", y = "Cluster")

micro_plot_filtered <- micro_plot + 
  scale_y_discrete(limits = c("MG", "AIF1+MG")) 
micro_plot_filtered

micro_plot  +
  guides(size = guide_legend(title = "Percent expressed", order = 2),
         color = guide_colorbar(title = "Avg. Expression (scaled)", order = 1)) +
  scale_size_continuous(breaks = c(0, 20, 40, 60),
                        limits = c(0, 100)) +
  scale_color_gradientn(
    colors = gradient,
    breaks = c(-1, 0, 1)
  ) +
  theme(
    legend.title = element_text(size = 14),
    legend.text = element_text(size = 12)
  )

saveRDS(micro.sub, file = "micro_sub.RDS")

#### Reincorporate labels back into main object ####

Idents(fb_seurat2) <- "cell_type_v20"

levels(fb_seurat2)

fb_seurat2 <- RenameIdents(fb_seurat2, c("MG_0" = "MG", "MG_1" = "MG", "MG_2" = "MG", "MG_3" = "AIF1+MG"))

DimPlot_scCustom(fb_seurat2, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

fb_seurat2$cell_type_v20 <- Idents(fb_seurat2)

saveRDS(fb_seurat2, file = "fb_seurat_v21.RDS")

micro_plot <- DotPlot_scCustom(fb_seurat, features = c( "CD68", "CD163", "AIF1", "C1QB", "C1QC",  "SALL1",
                                                        "CX3CR1", "P2RY12", "MRC1", "CSF1R", "TMEM119", "LYVE1", "GPR34", "CD80", "CD86", "IL1B", "IL6",
                                                        "TREM2", "APOE", "CTSD"), 
                               group.by = "cell_type_v20", remove_axis_titles = FALSE,  flip_axes = TRUE, x_lab_rotate = TRUE, colors_use = gradient) + labs(x = "Gene", y = "Cluster")

micro_plot_filtered <- micro_plot + 
  scale_y_discrete(limits = c("MG", "AIF1++ MG")) # Insert clusters to display here
micro_plot_filtered


micro_plot  +
  guides(size = guide_legend(title = "Percent expressed", order = 2),
         color = guide_colorbar(title = "Avg. Expression (scaled)", order = 1)) +
  scale_size_continuous(breaks = c(0, 20, 40, 60),
                        limits = c(0, 100)) +
  scale_color_gradientn(
    colors = gradient,
    breaks = c(-1, 0, 1)
  ) +
  theme(
    legend.title = element_text(size = 14),
    legend.text = element_text(size = 12)
  )



DimPlot_scCustom(micro.sub, group.by = "cell_type_v20", reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = TRUE), color_seed = 5)




# Specify your gene and the expression threshold
gene_of_interest <- "AIF1"
threshold <- 3

Idents(fb_seurat) <- "cell_type_v20"

high_aif1_cells <- WhichCells(
  object = fb_seurat,
  idents = c("MG",'AIF1+MG'),
  expression = AIF1 > 3
)

# Convert the original cluster identities to a character vector
fb_seurat$cell_type_v20 <- as.character(Idents(fb_seurat))

# Check the head of the new metadata column
head(fb_seurat$cell_type_v20)

# Assign the new cluster identity to the high_aif1 cells
fb_seurat$cell_type_v20[high_aif1_cells] <- "AIF1++ MG"

# Set the new_clusters column as the active identity
Idents(fb_seurat) <- "cell_type_v20"

DimPlot_scCustom(fb_seurat, group.by = "cell_type_v20", reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = TRUE), color_seed = 5)

fb_seurat<- RenameIdents(fb_seurat, "AIF1+MG" = "MG")

fb_seurat$cell_type_v20 <- Idents(fb_seurat)


avg <- AverageExpression(fb_seurat, features = "AIF1", layer = "data")

avg[['RNA']]["AIF1",]

avg <- AggregateExpression(fb_seurat, features = "AIF1")

fb_seurat <- RenameIdents(fb_seurat, "AIF1++ MG" = "AIF1++MG")
fb_seurat$cell_type_v20 <- Idents(fb_seurat)

order <- c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1", "GPC", "AS0", "AS1", "AS2", 'OPC0', 'OPC1', 'OPC2', "COP", 'OL1', 'OL2', "IPC", "DL-ExN", "UL-ExN", "AIF1++MG", "MG", "TRM", "EB", "EC", "PC", "EP")

fb_seurat$cell_type_v20 <- factor(
  x = fb_seurat$cell_type_v20,
  levels = order
)

saveRDS(fb_seurat, file = "fb_seurat_v21.RDS")


#### Generating subsets for UMAPs again ####
levels(fb_seurat)
clustersToSubset <- c("GPC", "OPC2", "OL2", "OPC1", "COP", "OPC0", "OL1")
oligo.sub <- subset(fb_seurat, idents = clustersToSubset)
saveRDS(oligo.sub, "oligo_sub.RDS")
DimPlot_scCustom(oligo.sub, reduction = "umap.harmony")
DimPlot_scCustom(oligo.sub, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = TRUE), color_seed = 5)

levels(fb_seurat)
clustersToSubset <- c("GPC", "AS0", "AS1", "AS2")
astro.sub <- subset(fb_seurat, idents = clustersToSubset)
saveRDS(astro.sub, 'astro_sub.RDS')
DimPlot_scCustom(astro.sub, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "polychrome", shuffle_pal = TRUE), color_seed = 5)

levels(fb_seurat)
clustersToSubset <- c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1")
neuro.sub <- subset(fb_seurat, idents = clustersToSubset)
saveRDS(neuro.sub, "neuro_sub.RDS")
DimPlot_scCustom(neuro.sub, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = TRUE), color_seed = 5)

levels(fb_seurat)
clustersToSubset <- c("IPC", "DL-ExN", "UL-ExN")
IPC.sub <- subset(fb_seurat, idents = clustersToSubset)
saveRDS(IPC.sub, "IPC_sub.RDS")
DimPlot_scCustom(IPC.sub, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "polychrome", shuffle_pal = TRUE), color_seed = 5)

levels(fb_seurat)
clustersToSubset <- c("AIF1++MG", "MG")
micro.sub <- subset(fb_seurat, idents = clustersToSubset)
saveRDS(micro.sub, "micro_sub.RDS")
DimPlot_scCustom(micro.sub, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "polychrome", shuffle_pal = TRUE), color_seed = 5)



#### Re-exporting for trajectory analysis for analysis, for the last time ####

# Convert to anndata for scVelo analysis using scCustomize
as.anndata(
  x = fb_seurat,
  file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain",
  file_name = "fb_adata_3.h5ad"
)

# Convert micro.sub to anndata for scVelo analysis using scCustomize
as.anndata(
  x = micro.sub,
  file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain",
  file_name = "micro_adata_2.h5ad"
)

# Convert neuron.sub to anndata for scVelo analysis using scCustomize
as.anndata(
  x = neuro.sub,
  file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain",
  file_name = "neuro_adata.h5ad"
)

# Convert oligo.sub to anndata for scVelo analysis using scCustomize
as.anndata(
  x = oligo.sub,
  file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain",
  file_name = "oligo_adata.h5ad"
)

# Convert astro.sub to anndata for scVelo analysis using scCustomize
as.anndata(
  x = astro.sub,
  file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain",
  file_name = "astro_adata.h5ad"
)

# Convert IPC.sub to anndata for scVelo analysis using scCustomize
as.anndata(
  x = IPC.sub,
  file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain",
  file_name = "IPC_adata.h5ad"
)


#### Change names on IPCs ####

fb_seurat <- RenameIdents(fb_seurat, c("UL-ExN" = "ExN2", "DL-ExN" = "ExN1", "IPC" = "ExN0", "GPC" = "ExN3"))
levels(fb_seurat)

fb_seurat <- RenameIdents(fb_seurat, "AIF1++MG" = "MG")
DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, pt.size = 0.5, repel = TRUE, label.box = TRUE)

#### Get top markers for ExN3 ####
fb_seurat <- PrepSCTFindMarkers(fb_seurat)

all_markers <- FindAllMarkers(object = fb_seurat) %>%
  Add_Pct_Diff() 

not_all_markers <- all_markers[grepl("ExN3", all_markers$cluster), ]

top_markers <- Extract_Top_Markers(marker_dataframe = not_all_markers, num_genes = 20, named_vector = FALSE,
                                   make_unique = TRUE, rank_by = "avg_log2FC")
DotPlot_scCustom(fb_seurat, top_markers, x_lab_rotate = TRUE, flip_axes = TRUE)

FeaturePlot_scCustom(fb_seurat, reduction = "umap.harmony", c('SLC17A7', "SLC17A6", 'GRIA1', 'GRIA2', 'GRIA3', 'GRIA4', 'GRIN1', 'GRIN2A', 'GRIN2B', 'GRIN2C', 'GRIN2D'))
DotPlot_scCustom(fb_seurat, c('SLC17A7', "SLC17A6", 'GRIA1', 'GRIA2', 'GRIA3', 'GRIA4', 'GRIN1', 'GRIN2A', 'GRIN2B', 'GRIN2C', 'GRIN2D'), flip_axes = TRUE, x_lab_rotate = TRUE)

FeaturePlot_scCustom(fb_seurat, reduction = "umap.harmony", c('SATB2', 'TBR1', 'BCL11B', 'CTIP2'))
DotPlot_scCustom(fb_seurat, c('ENSMMUG00000052254', 'SLC17A7', 'CAMK2A', "NRGN", "DCX", 'STMN2'), x_lab_rotate = TRUE, flip_axes = TRUE)

FeaturePlot_scCustom(fb_seurat, reduction = "umap.harmony", "BCAS1")
DotPlot_scCustom(fb_seurat, "BCAS1")

fb_seurat$cell_type_v20 <- Idents(fb_seurat)
fb_seurat$cell_type_v20 <- Idents(fb_seurat)
saveRDS(fb_seurat, file = "fb_seurat_v21.RDS")


order <- c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1", "ExN3", "AS0", "AS1", "AS2", 'OPC0', 'OPC1', 'OPC2', "COP", 'OL1', 'OL2', "ExN0", "ExN1", "ExN2", "MG", "TRM", "EB", "EC", "PC", "EP")

fb_seurat$cell_type_v20 <- factor(
  x = fb_seurat$cell_type_v20,
  levels = order
)

Idents(fb_seurat) <- "cell_type_v20"

DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = FALSE), color_seed = 5)

saveRDS(fb_seurat, file = "fb_seurat_v21.RDS")


#### Generating subsets for UMAPs again again ####
levels(fb_seurat)
clustersToSubset <- c("OPC2", "OL2", "OPC1", "COP", "OPC0", "OL1")
oligo.sub <- subset(fb_seurat, idents = clustersToSubset)
saveRDS(oligo.sub, "oligo_sub.RDS")
DimPlot_scCustom(oligo.sub, reduction = "umap.harmony")
DimPlot_scCustom(oligo.sub, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = TRUE), color_seed = 5)

# Convert oligo.sub to anndata for scVelo analysis using scCustomize
as.anndata(
  x = oligo.sub,
  file_path = "/mmfs1/gscratch/kawaldorflab/jcorn427/fet_brain",
  file_name = "oligo_adata.h5ad"
)

FeaturePlot_scCustom(fb_seurat, reduction = "umap.harmony", c('PDGFRA', 'CSPG4', 'OLIG1', 'OLIG2', 'GPR17', 'VIM', 'HOPX', 'SOX9', 'FABP7', 'ID3'), num_columns = 3)

FeaturePlot_scCustom(fb_seurat, reduction = "umap.harmony", c('PDGFRA'), na_cutoff = 3)

FeaturePlot_scCustom(fb_seurat, reduction = "umap.harmony", c('CSPG4'))

FeaturePlot_scCustom(fb_seurat, reduction = "umap.harmony", c('OLIG2'), na_cutoff = 3)

FeaturePlot_scCustom(fb_seurat, reduction = "umap.harmony", c('SOX10'), na_cutoff = 2)

FeaturePlot_scCustom(fb_seurat, reduction = "umap.harmony", c('ENSMMUG00000056728'), na_cutoff = 2)

FeaturePlot_scCustom(fb_seurat, reduction = "umap.harmony", c('GPR17'), na_cutoff = 2)

FeaturePlot_scCustom(fb_seurat, reduction = "umap.harmony", c('PCDH15'), na_cutoff = 2)

FeaturePlot_scCustom(fb_seurat, reduction = "umap.harmony", c('SOX9', 'HOPX', 'FABP7', 'ID3'), na_cutoff = 2)

FeaturePlot_scCustom(fb_seurat, reduction = "umap.harmony", c('TNR', 'LHFPL3'), na_cutoff = 2)

not_all_markers <- all_markers[grepl(paste(c("OPC1", "OPC2"), collapse = "|"), all_markers$cluster), ]

top_markers <- Extract_Top_Markers(marker_dataframe = not_all_markers, num_genes = 7, named_vector = FALSE,
                                   make_unique = TRUE, rank_by = "avg_log2FC")
DotPlot_scCustom(fb_seurat, top_markers, x_lab_rotate = TRUE, flip_axes = TRUE)

#### Seurat to Monocle3 ####

cds <- as.cell_data_set(fb_seurat)

# Get the coordinates for your UMAP.HARMONY
harmony_coords <- reducedDims(cds)[["UMAP.HARMONY"]]

# Overwrite the default "UMAP" coordinates with your harmony-integrated coordinates
reducedDims(cds)[["UMAP"]] <- harmony_coords


plot_cells(cds, color_cells_by="cell_type_v20", show_trajectory_graph=FALSE, reduction_method = "UMAP")
plot_cells()

# Now run the rest of the Monocle3 workflow
cds <- cluster_cells(cds, reduction_method = "UMAP")
cds <- learn_graph(cds, use_partition = TRUE)
cds <- order_cells(cds)

# Plot the trajectory and pseudotime on the Harmony UMAP
plot_cells(cds,
           #color_cells_by = "pseudotime",
           label_cell_groups = FALSE,
           label_leaves = FALSE,
           label_branch_points = FALSE)

plot_cells(cds,
           color_cells_by = "cell_type_v20",
           label_principal_points = TRUE)



cds <- order_cells(cds, root_pr_nodes = c("Y_157", "Y_292", "Y_211", "Y_27", "Y_3", "Y_402", "Y_296", "Y_338", "Y_373"))

plot_cells(cds,
           color_cells_by = "pseudotime",
           show_trajectory_graph = TRUE,
           label_cell_groups = FALSE, # Set to TRUE to label cell groups
           label_leaves = TRUE,
           label_branch_points = TRUE,
           graph_label_size = 1.5)


FeaturePlot_scCustom(fb_seurat, reduction = "umap.harmony", c('GATA1', 'KLF1', 'TAL1', 'SCL', 'ZBTB7A', 'LRF', "GYPA"))
FeaturePlot_scCustom(fb_seurat, reduction = "umap.harmony", c('HBA1', 'HBA', 'HBM', 'ALAS2', 'HBG', 'ENSMMUG00000044429'))
FeaturePlot_scCustom(fb_seurat, reduction = "umap.harmony", c('PF4', 'ITGA2B', 'GP9'))
FeaturePlot_scCustom(fb_seurat, reduction = "umap.harmony", c('CD3D'))



DimPlot_scCustom(fb_seurat2, reduction = "umap.harmony", group.by = "cell_type_v22")

fb_seurat2 <- FindSubCluster(object = fb_seurat, cluster = "TRM", graph.name = 'SCT_snn', subcluster.name = "cell_type_v21", resolution = 0.25)

Idents(fb_seurat2) <- "cell_type_v21"

fb_seurat2 <- FindSubCluster(object = fb_seurat2, cluster = "EB", graph.name = 'SCT_snn', subcluster.name = "cell_type_v22", resolution = 0.25)

Idents(fb_seurat2) <- "cell_type_v22"

fb_seurat <- PrepSCTFindMarkers(fb_seurat2)

all_markers <- FindAllMarkers(object = fb_seurat) %>%
  Add_Pct_Diff() 

not_all_markers <- all_markers[grepl(paste(c("EB_0", "EB_1", "TRM_0", "TRM_1"), collapse = "|"), all_markers$cluster), ]

top_markers <- Extract_Top_Markers(marker_dataframe = not_all_markers, num_genes = 20, named_vector = FALSE,
                                   make_unique = TRUE, rank_by = "avg_log2FC")
DotPlot_scCustom(fb_seurat, top_markers, x_lab_rotate = TRUE, flip_axes = TRUE)

DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, label.box = TRUE, repel = TRUE)

DotPlot_scCustom(fb_seurat, "CD14", x_lab_rotate = TRUE, flip_axes = TRUE)

fb_seurat <- RenameIdents(fb_seurat, "TRM_0" = "Mono")


not_all_markers <- all_markers[grepl("EB_0", all_markers$cluster), ]

top_markers <- Extract_Top_Markers(marker_dataframe = not_all_markers, num_genes = 20, named_vector = FALSE,
                                   make_unique = TRUE, rank_by = "avg_log2FC")
DotPlot_scCustom(fb_seurat, top_markers, x_lab_rotate = TRUE, flip_axes = TRUE)

DotPlot_scCustom(fb_seurat, c("CD4", "CD8A"), x_lab_rotate = TRUE, flip_axes = TRUE)

fb_seurat <- RenameIdents(fb_seurat, "TRM_1" = "CD4+T")

DotPlot_scCustom(fb_seurat, c('COL1A1', 'COL1A2', 'COL3A1', 'FN1', 'FBN1', 'VCAN', 'POSTN', 'MMP2', 'MMP3', 'MMP14', 'TIMP1', 'LOX'), x_lab_rotate = TRUE, flip_axes = TRUE)

fb_seurat <- RenameIdents(fb_seurat, c("EB_0" = "EB1", "EB_1" = "EB0"))


DotPlot_scCustom(fb_seurat, c('COL1A1', 'COL1A2', 'COL3A1', 'FN1', 'FBN1', 'VCAN', 'POSTN', 'MMP2', 'MMP3', 'MMP14', 'TIMP1', 'LOX'), x_lab_rotate = TRUE, flip_axes = TRUE)

FeaturePlot_scCustom(fb_seurat, features = c('ACTA2', 'MYH11', 'CNN1', 'TAGLN', 'MYLK', 'CALD1'), reduction = "umap.harmony")

FeaturePlot_scCustom(fb_seurat, features = c('MKI67', 'PECAM1', "VWF"), reduction = "umap.harmony")

FeaturePlot_scCustom(fb_seurat, features = "FLT1", reduction = "umap.harmony", na_cutoff = 3)

FeaturePlot_scCustom(fb_seurat, features = c('FOXJ1', 'DNAH5', 'DNAH9', 'RSPH1', 'RSPH9', 'TUBB4B', 'S100B', 'SPP1', 'GFAP', 'VIM'), reduction = "umap.harmony")

FeaturePlot_scCustom(fb_seurat, features = c("S100B"), reduction = "umap.harmony")


DimPlot_scCustom(fb_seurat, reduction = "umap.harmony")

fb_seurat$cell_type_v22 <- Idents(fb_seurat)

saveRDS(fb_seurat, "fb_seurat_v22.RDS")

fb_seurat2 <- FindSubCluster(object = fb_seurat, cluster = "ExN0", graph.name = 'SCT_snn', subcluster.name = "cell_type_v23", resolution = 0.25)

DimPlot_scCustom(fb_seurat2, reduction = "umap.harmony", group.by = "cell_type_v23", label = TRUE)
levels(fb_seurat2)
Idents(fb_seurat2) <- "cell_type_v23"

fb_seurat2 <- FindSubCluster(object = fb_seurat2, cluster = "ExN2", graph.name = 'SCT_snn', subcluster.name = "cell_type_v24", resolution = 0.01)

DimPlot_scCustom(fb_seurat2, reduction = "umap.harmony", group.by = "cell_type_v24", label = TRUE)
levels(fb_seurat2)
Idents(fb_seurat2) <- "cell_type_v24"

fb_seurat2 <- RenameIdents(fb_seurat2, c("ExN2_0" = "ExN2", "ExN2_1" = "ExN2", "ExN2_2" = "ExN3", "ExN0_0" = "ExN0", 'ExN0_1' = "CGE0"))

DimPlot_scCustom(fb_seurat2, reduction = "umap.harmony", label = TRUE)

DimPlot_scCustom(fb_seurat2, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = FALSE), color_seed = 5)

fb_seurat <- fb_seurat2

fb_seurat$cell_type_v24 <- Idents(fb_seurat)

Idents(fb_seurat) <- "cell_type_v24"

order <- c("ExN0", "ExN1", "ExN2", "ExN3", "vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1", "MG", "CD4+T", "Mono", 'OPC0', 'OPC1', 'OPC2', "COP", 'OL1', 'OL2', "AS0", "AS1", "AS2", "EB0", "EB1", "EC", "PC", "EP")

fb_seurat$cell_type_v25 <- factor(
  x = fb_seurat$cell_type_v24,
  levels = order
)

Idents(fb_seurat) <- "cell_type_v25"

DimPlot_scCustom(fb_seurat, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = FALSE), color_seed = 5)

saveRDS(fb_seurat, file = "fb_seurat_v25.RDS")






#### Seurat to Monocle3 ####

cds <- as.cell_data_set(fb_seurat)

# Get the coordinates for your UMAP.HARMONY
harmony_coords <- reducedDims(cds)[["UMAP.HARMONY"]]

# Overwrite the default "UMAP" coordinates with your harmony-integrated coordinates
reducedDims(cds)[["UMAP"]] <- harmony_coords


plot_cells(cds, color_cells_by="cell_type_v25", show_trajectory_graph=FALSE, reduction_method = "UMAP")


# Now run the rest of the Monocle3 workflow
cds <- cluster_cells(cds, reduction_method = "UMAP")
cds <- learn_graph(cds, use_partition = TRUE)
cds <- order_cells(cds)

# Plot the trajectory and pseudotime on the Harmony UMAP
plot_cells(cds,
           color_cells_by = "pseudotime",
           label_cell_groups = FALSE,
           label_leaves = FALSE,
           label_branch_points = FALSE)

plot_cells(cds,
           color_cells_by = "cell_type_v25",
           label_principal_points = FALSE, group_label_size = 3)



cds <- order_cells(cds, root_pr_nodes = c("Y_157", "Y_292", "Y_211", "Y_27", "Y_3", "Y_402", "Y_296", "Y_338", "Y_373"))

plot_cells(cds,
           color_cells_by = "pseudotime",
           show_trajectory_graph = TRUE,
           label_cell_groups = TRUE, # Set to TRUE to label cell groups
           label_leaves = TRUE,
           label_branch_points = TRUE,
           graph_label_size = 1.5)



#### Generating subsets for UMAPs again ####
levels(fb_seurat)
clustersToSubset <- c("OPC2", "OL2", "OPC1", "COP", "OPC0", "OL1")
oligo.sub <- subset(fb_seurat, idents = clustersToSubset)
saveRDS(oligo.sub, "oligo_sub.RDS")
DimPlot_scCustom(oligo.sub, reduction = "umap.harmony")
DimPlot_scCustom(oligo.sub, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = TRUE), color_seed = 5)

levels(fb_seurat)
clustersToSubset <- c("AS0", "AS1", "AS2")
astro.sub <- subset(fb_seurat, idents = clustersToSubset)
saveRDS(astro.sub, 'astro_sub.RDS')
DimPlot_scCustom(astro.sub, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "polychrome", shuffle_pal = TRUE), color_seed = 5)

levels(fb_seurat)
clustersToSubset <- c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1")
neuro.sub <- subset(fb_seurat, idents = clustersToSubset)
saveRDS(neuro.sub, "neuro_sub.RDS")
DimPlot_scCustom(neuro.sub, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = TRUE), color_seed = 5)

levels(fb_seurat)
clustersToSubset <- c("ExN0", "ExN1", "ExN2", "ExN3")
ExN.sub <- subset(fb_seurat, idents = clustersToSubset)
saveRDS(ExN.sub, "ExN_sub.RDS")
DimPlot_scCustom(ExN.sub, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "polychrome", shuffle_pal = TRUE), color_seed = 5)

levels(fb_seurat)
clustersToSubset <- c("MG")
micro.sub <- subset(fb_seurat, idents = clustersToSubset)
saveRDS(micro.sub, "micro_sub.RDS")
DimPlot_scCustom(micro.sub, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "polychrome", shuffle_pal = TRUE), color_seed = 5)

levels(fb_seurat)
clustersToSubset <- c("OPC0", "AS0", "ExN3")
odd.sub <- subset(fb_seurat, idents = clustersToSubset)
saveRDS(odd.sub, "odd_sub.RDS")
DimPlot_scCustom(odd.sub, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "polychrome", shuffle_pal = TRUE), color_seed = 5)

#### Oligo Monocle3 ####

cds <- as.cell_data_set(oligo.sub)

# Get the coordinates for your UMAP.HARMONY
harmony_coords <- reducedDims(cds)[["UMAP.HARMONY"]]

# Overwrite the default "UMAP" coordinates with your harmony-integrated coordinates
reducedDims(cds)[["UMAP"]] <- harmony_coords


plot_cells(cds, color_cells_by="cell_type_v25", show_trajectory_graph=FALSE, reduction_method = "UMAP")


# Now run the rest of the Monocle3 workflow
cds <- cluster_cells(cds, reduction_method = "UMAP")
cds <- learn_graph(cds, use_partition = TRUE)
cds <- order_cells(cds)

# Plot the trajectory and pseudotime on the Harmony UMAP
plot_cells(cds,
           color_cells_by = "pseudotime",
           label_cell_groups = FALSE,
           label_leaves = FALSE,
           label_branch_points = FALSE)

plot_cells(cds,
           color_cells_by = "cell_type_v25",
           label_principal_points = TRUE, group_label_size = 3)



cds <- order_cells(cds, root_pr_nodes = c("Y_26", "Y_71", "Y_80"))

plot_cells(cds,
           color_cells_by = "pseudotime",
           show_trajectory_graph = TRUE,
           label_cell_groups = TRUE, # Set to TRUE to label cell groups
           label_leaves = TRUE,
           label_branch_points = TRUE,
           graph_label_size = 1.5)


# Get cell counts per cluster per sample
cell_counts <- table(fb_seurat@meta.data$cell_type_v25, fb_seurat@meta.data$sample)
cell_counts <- as.data.frame(cell_counts)

ggplot(data=cell_counts, aes(x=Var1, y=Freq, fill = Var2)) +
  geom_bar(stat="identity", position="fill") + 
  labs(title = "Cell Proportion Per Cell Type Per Sample", x= "Cell Type", y= "Cell Counts") + 
  guides(fill = guide_legend(title = "Sample",reverse=FALSE)) +
  #scale_fill_discrete(labels = c("CV", "Decidua", "Meminoc")) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

Plot_Density_Custom(fb_seurat, features = c("EOMES", "TBR1", "SATB2"), joint = TRUE, reduction = "umap.harmony")


#### Generate intersect plot of genes of interest shuffled ####

umap_coords <- as.data.frame(ExN_sub@reductions$umap.harmony@cell.embeddings)

# Ensure the genes exist in your object before proceeding
genes_to_plot <- c("TBR1", "EOMES", "SATB2")
if (!all(genes_to_plot %in% rownames(fb_seurat))) {
  stop("One or more genes not found in the Seurat object.")
}

# Fetch expression data. `FetchData` is the safest way to get expression.
expression_data <- FetchData(object = ExN_sub, vars = genes_to_plot)

# Data for the background layer (all cells)
background_data <- umap_coords

# Prepare data for the shuffled foreground layers (expressing cells only)
plot_data <- cbind(umap_coords, expression_data)

long_data <- plot_data %>%
  pivot_longer(
    cols = all_of(genes_to_plot),
    names_to = "gene",
    values_to = "expression"
  ) %>%
  # Filter for cells that express the gene (expression > 0)
  filter(expression > 0)

# Randomly shuffle the rows of the combined data frame
shuffled_data <- long_data %>%
  sample_frac(1)

ggplot() +
  # 1. Plot all cells as a light grey background layer
  geom_point(data = background_data,
             aes(x = umapharmony_1, y = umapharmony_2),
             color = "lightgrey",
             size = 0.5,
             alpha = 0.5) +
  
  # 2. Plot the shuffled, expressing cells on top
  geom_point(data = shuffled_data,
             aes(x = umapharmony_1, y = umapharmony_2, color = gene),
             alpha = 0.5,
             size = 0.8) +
  
  # Set manual colors and customize the theme
  scale_color_manual(values = c("TBR1" = "darkgoldenrod1",
                                "EOMES" = "darkorchid4",
                                "SATB2" = "green")) +
  labs(title = "EOMES, SATB2, TBR1",
       color = "Gene Expression") +
  theme_minimal() +
  theme(
    panel.grid = element_blank(), # Remove grid lines
    axis.line = element_line(color = "black") # Add axis lines back in
  )





#### Lasso cells of interest ####

# Generate the UMAP plot
plot <- DimPlot(object = ExN_sub, reduction = "umap.harmony")

# Pass the plot to CellSelector()
# The R console will prompt you to select points on the plot
cells.located <- CellSelector(plot = plot)
cells.located2 <- CellSelector(plot = plot)
# Create a new identity class for the selected cells
ExN_sub <- SetIdent(ExN_sub, cells = cells.located, value = 'selected_cells')
ExN_sub <- SetIdent(ExN_sub, cells = cells.located2, value = 'selected_cells2')

ExN_sub <- RenameIdents(
  object = ExN_sub,
  'selected_cells' = "CombinedGroup",
  'selected_cells2' = "CombinedGroup"
)
# Verify the selection with a new plot
DimPlot(ExN_sub, reduction = "umap.harmony", group.by = 'ident')

ExN_sub$lasso <- Idents(ExN_sub)

saveRDS(ExN_sub, 'ExN_sub.RDS')

combined_group_subset <- subset(ExN_sub, idents = "CombinedGroup")

DimPlot(ExN_sub, reduction = "umap.harmony", group.by = 'lasso')





#### Use new subset for gene expression of interest ####
umap_coords <- as.data.frame(combined_group_subset@reductions$umap.harmony@cell.embeddings)

# Ensure the genes exist in your object before proceeding
genes_to_plot <- c("TBR1", "EOMES", "SATB2")
if (!all(genes_to_plot %in% rownames(fb_seurat))) {
  stop("One or more genes not found in the Seurat object.")
}

# Fetch expression data. `FetchData` is the safest way to get expression.
expression_data <- FetchData(object = combined_group_subset, vars = genes_to_plot)

# Data for the background layer (all cells)
background_data <- umap_coords

# Prepare data for the shuffled foreground layers (expressing cells only)
plot_data <- cbind(umap_coords, expression_data)

long_data <- plot_data %>%
  pivot_longer(
    cols = all_of(genes_to_plot),
    names_to = "gene",
    values_to = "expression"
  ) %>%
  # Filter for cells that express the gene (expression > 0)
  filter(expression > 0)

# Randomly shuffle the rows of the combined data frame
shuffled_data <- long_data %>%
  sample_frac(1)

ggplot() +
  # 1. Plot all cells as a light grey background layer
  geom_point(data = background_data,
             aes(x = umapharmony_1, y = umapharmony_2),
             color = "lightgrey",
             size = 0.5,
             alpha = 0.5) +
  
  # 2. Plot the shuffled, expressing cells on top
  geom_point(data = shuffled_data,
             aes(x = umapharmony_1, y = umapharmony_2, color = gene),
             alpha = 0.5,
             size = 0.8) +
  
  # Set manual colors and customize the theme
  scale_color_manual(values = c("TBR1" = "green",
                                "EOMES" = "blue",
                                "SATB2" = "orange")) +
  # Use guides() to override the size of the points in the legend
  guides(color = guide_legend(override.aes = list(size = 4))) +
  labs(title = "EOMES, SATB2, TBR1",
       color = "Gene Expression") +
  theme_minimal() +
  theme(
    panel.grid = element_blank(), # Remove grid lines
    axis.line = element_line(color = "black"), # Add axis lines back in
    # Increase legend title and text size
    legend.title = element_text(size = 14), 
    legend.text = element_text(size = 12)
  )

umap_coords <- as.data.frame(combined_group_subset@reductions$umap.harmony@cell.embeddings)

# Ensure the genes exist in your object before proceeding
genes_to_plot <- c("POU3F2", "SATB2")
if (!all(genes_to_plot %in% rownames(fb_seurat))) {
  stop("One or more genes not found in the Seurat object.")
}

# Fetch expression data. `FetchData` is the safest way to get expression.
expression_data <- FetchData(object = combined_group_subset, vars = genes_to_plot)

# Data for the background layer (all cells)
background_data <- umap_coords

# Prepare data for the shuffled foreground layers (expressing cells only)
plot_data <- cbind(umap_coords, expression_data)

long_data <- plot_data %>%
  pivot_longer(
    cols = all_of(genes_to_plot),
    names_to = "gene",
    values_to = "expression"
  ) %>%
  # Filter for cells that express the gene (expression > 0)
  filter(expression > 0)

# Randomly shuffle the rows of the combined data frame
shuffled_data <- long_data %>%
  sample_frac(1)

ggplot() +
  # 1. Plot all cells as a light grey background layer
  geom_point(data = background_data,
             aes(x = umapharmony_1, y = umapharmony_2),
             color = "lightgrey",
             size = 0.5,
             alpha = 0.5) +
  
  # 2. Plot the shuffled, expressing cells on top
  geom_point(data = shuffled_data,
             aes(x = umapharmony_1, y = umapharmony_2, color = gene),
             alpha = 0.5,
             size = 0.8) +
  
  # Set manual colors and customize the theme
  scale_color_manual(values = c("POU3F2" = "purple",
                                "SATB2" = "green")) +
  labs(title = "POU3F2, SATB2",
       color = "Gene Expression") +
  theme_minimal() +
  theme(
    panel.grid = element_blank(), # Remove grid lines
    axis.line = element_line(color = "black") # Add axis lines back in
  )





umap_coords <- as.data.frame(combined_group_subset@reductions$umap.harmony@cell.embeddings)

# Ensure the genes exist in your object before proceeding
genes_to_plot <- c("SLC17A6", "SLC17A7")
if (!all(genes_to_plot %in% rownames(fb_seurat))) {
  stop("One or more genes not found in the Seurat object.")
}

# Fetch expression data. `FetchData` is the safest way to get expression.
expression_data <- FetchData(object = combined_group_subset, vars = genes_to_plot)

# Data for the background layer (all cells)
background_data <- umap_coords

# Prepare data for the shuffled foreground layers (expressing cells only)
plot_data <- cbind(umap_coords, expression_data)

long_data <- plot_data %>%
  pivot_longer(
    cols = all_of(genes_to_plot),
    names_to = "gene",
    values_to = "expression"
  ) %>%
  # Filter for cells that express the gene (expression > 0)
  filter(expression > 0)

# Randomly shuffle the rows of the combined data frame
shuffled_data <- long_data %>%
  sample_frac(1)

ggplot() +
  # 1. Plot all cells as a light grey background layer
  geom_point(data = background_data,
             aes(x = umapharmony_1, y = umapharmony_2),
             color = "lightgrey",
             size = 0.5,
             alpha = 0.5) +
  
  # 2. Plot the shuffled, expressing cells on top
  geom_point(data = shuffled_data,
             aes(x = umapharmony_1, y = umapharmony_2, color = gene),
             alpha = 0.5,
             size = 0.8) +
  
  # Set manual colors and customize the theme
  scale_color_manual(values = c(
                                "SLC17A6" = "orange",
                                "SLC17A7" = "blue")) +
  # Use guides() to override the size of the points in the legend
  guides(color = guide_legend(override.aes = list(size = 4))) +
  labs(title = "SLC17A6, SLC17A7",
       color = "Gene Expression") +
  theme_minimal() +
  theme(
    panel.grid = element_blank(), # Remove grid lines
    axis.line = element_line(color = "black"), # Add axis lines back in
    # Increase legend title and text size
    legend.title = element_text(size = 14), 
    legend.text = element_text(size = 12)
  )



# Correctly create the new metadata column using as.character()
ExN_sub$cell_type_updated <- ifelse(
  test = ExN_sub$lasso == "CombinedGroup",
  yes = as.character(ExN_sub$cell_type_v25),
  no = "other"
)

# You can convert the column back to a factor if needed for plotting
# R will automatically convert character vectors to factors when they are used for grouping
# so this is often not required for standard plotting functions
ExN_sub$cell_type_updated <- as.factor(ExN_sub$cell_type_updated)

# Check the new assignments
table(ExN_sub$cell_type_updated)

# Now, plot with the new, correctly labeled column
DimPlot_scCustom(ExN_sub, group.by = "cell_type_updated", reduction = "umap.harmony")

Idents(ExN_sub) <- "cell_type_updated"
subset_to_plot <- subset(ExN_sub, idents = "other", invert = TRUE)

DimPlot_scCustom(subset_to_plot, group.by = "cell_type_updated", reduction = "umap.harmony")
DimPlot_scCustom(subset_to_plot, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "polychrome", shuffle_pal = TRUE), color_seed = 5)


#### CGE/MGE work ####

# Generate the UMAP plot
plot <- DimPlot(object = neuro_sub, reduction = "umap.harmony") # replace neuro_sub with subset of interest

# Pass the plot to CellSelector()
# The R console will prompt you to select points on the plot
cells.located <- CellSelector(plot = plot)
cells.located2 <- CellSelector(plot = plot)
# Create a new identity class for the selected cells
neuro_sub <- SetIdent(neuro_sub, cells = cells.located, value = 'selected_cells')
neuro_sub <- SetIdent(neuro_sub, cells = cells.located2, value = 'selected_cells2')

neuro_sub <- RenameIdents(
  object = neuro_sub,
  'selected_cells' = "CombinedGroup",
  'selected_cells2' = "CombinedGroup"
)
# Verify the selection with a new plot
DimPlot(neuro_sub, reduction = "umap.harmony", group.by = 'ident')

neuro_sub$lasso <- Idents(neuro_sub)

# Correctly create the new metadata column using as.character()
neuro_sub$cell_type_updated <- ifelse(
  test = neuro_sub$lasso == "CombinedGroup",
  yes = as.character(neuro_sub$cell_type_v25),
  no = "other"
)

# You can convert the column back to a factor if needed for plotting
# R will automatically convert character vectors to factors when they are used for grouping
# so this is often not required for standard plotting functions
neuro_sub$cell_type_updated <- as.factor(neuro_sub$cell_type_updated)

# Check the new assignments
table(neuro_sub$cell_type_updated)

# Now, plot with the new, correctly labeled column
DimPlot_scCustom(neuro_sub, group.by = "cell_type_updated", reduction = "umap.harmony")

Idents(neuro_sub) <- "cell_type_updated"
subset_to_plot <- subset(neuro_sub, idents = "other", invert = TRUE)

DimPlot_scCustom(subset_to_plot, group.by = "cell_type_updated", reduction = "umap.harmony")
DimPlot_scCustom(subset_to_plot, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = TRUE, seed = 2))

saveRDS(neuro_sub, 'neuro_sub.RDS')

### Plot genes of interest ###
Idents(neuro_sub) <- "cell_type_updated"
subset_to_plot <- subset(neuro_sub, idents = "other", invert = TRUE) # replace neur_sub with subset of interest
umap_coords <- as.data.frame(subset_to_plot@reductions$umap.harmony@cell.embeddings)

# Ensure the genes exist in your object before proceeding
genes_to_plot <- c("SLC17A6", "SLC17A7") # replace with your genes of interest
if (!all(genes_to_plot %in% rownames(fb_seurat))) {
  stop("One or more genes not found in the Seurat object.")
}

# Fetch expression data. `FetchData` is the safest way to get expression.
expression_data <- FetchData(object = subset_to_plot, vars = genes_to_plot)

# Data for the background layer (all cells)
background_data <- umap_coords

# Prepare data for the shuffled foreground layers (expressing cells only)
plot_data <- cbind(umap_coords, expression_data)

long_data <- plot_data %>%
  pivot_longer(
    cols = all_of(genes_to_plot),
    names_to = "gene",
    values_to = "expression"
  ) %>%
  # Filter for cells that express the gene (expression > 0)
  filter(expression > 0)

# Randomly shuffle the rows of the combined data frame
shuffled_data <- long_data %>%
  sample_frac(1)

ggplot() +
  # 1. Plot all cells as a light grey background layer
  geom_point(data = background_data,
             aes(x = umapharmony_1, y = umapharmony_2),
             color = "lightgrey",
             size = 0.5,
             alpha = 0.5) +
  
  # 2. Plot the shuffled, expressing cells on top
  geom_point(data = shuffled_data,
             aes(x = umapharmony_1, y = umapharmony_2, color = gene),
             alpha = 0.5,
             size = 0.8) +
  
  # Set manual colors and customize the theme, also, remember to replace genes with those you are interested in
  scale_color_manual(values = c(
    "SLC17A6" = "orange",
    "SLC17A7" = "blue")) +
  # Use guides() to override the size of the points in the legend
  guides(color = guide_legend(override.aes = list(size = 4))) +
  labs(title = "SLC17A6, SLC17A7",
       color = "Gene Expression") +
  theme_minimal() +
  theme(
    panel.grid = element_blank(), # Remove grid lines
    axis.line = element_line(color = "black"), # Add axis lines back in
    # Increase legend title and text size
    legend.title = element_text(size = 14), 
    legend.text = element_text(size = 12)
  )


#### Oligo work ####

# Generate the UMAP plot
plot <- DimPlot(object = oligo_sub, reduction = "umap.harmony") # replace neuro_sub with subset of interest

# Pass the plot to CellSelector()
# The R console will prompt you to select points on the plot
cells.located <- CellSelector(plot = plot)
cells.located2 <- CellSelector(plot = plot)
# Create a new identity class for the selected cells
oligo_sub <- SetIdent(oligo_sub, cells = cells.located, value = 'selected_cells')
oligo_sub <- SetIdent(oligo_sub, cells = cells.located2, value = 'selected_cells2')

oligo_sub <- RenameIdents(
  object = oligo_sub,
  'selected_cells' = "CombinedGroup",
  'selected_cells2' = "CombinedGroup"
)
# Verify the selection with a new plot
DimPlot(oligo_sub, reduction = "umap.harmony", group.by = 'ident')

oligo_sub$lasso <- Idents(oligo_sub)

# Correctly create the new metadata column using as.character()
oligo_sub$cell_type_updated <- ifelse(
  test = oligo_sub$lasso == "CombinedGroup",
  yes = as.character(oligo_sub$cell_type_v25),
  no = "other"
)

# You can convert the column back to a factor if needed for plotting
# R will automatically convert character vectors to factors when they are used for grouping
# so this is often not required for standard plotting functions
oligo_sub$cell_type_updated <- as.factor(oligo_sub$cell_type_updated)

# Check the new assignments
table(oligo_sub$cell_type_updated)

# Now, plot with the new, correctly labeled column
DimPlot_scCustom(oligo_sub, group.by = "cell_type_updated", reduction = "umap.harmony")

Idents(oligo_sub) <- "cell_type_updated"
subset_to_plot <- subset(oligo_sub, idents = "other", invert = TRUE)

DimPlot_scCustom(subset_to_plot, group.by = "cell_type_updated", reduction = "umap.harmony")
DimPlot_scCustom(subset_to_plot, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = TRUE), color_seed = 5)

saveRDS(oligo_sub, 'oligo_sub.RDS')

### Plot genes of interest ###
Idents(oligo_sub) <- "cell_type_updated"
subset_to_plot <- subset(oligo_sub, idents = "other", invert = TRUE) # replace neur_sub with subset of interest
umap_coords <- as.data.frame(subset_to_plot@reductions$umap.harmony@cell.embeddings)

# Ensure the genes exist in your object before proceeding
genes_to_plot <- c("SLC17A6", "SLC17A7") # replace with your genes of interest
if (!all(genes_to_plot %in% rownames(fb_seurat))) {
  stop("One or more genes not found in the Seurat object.")
}

# Fetch expression data. `FetchData` is the safest way to get expression.
expression_data <- FetchData(object = subset_to_plot, vars = genes_to_plot)

# Data for the background layer (all cells)
background_data <- umap_coords

# Prepare data for the shuffled foreground layers (expressing cells only)
plot_data <- cbind(umap_coords, expression_data)

long_data <- plot_data %>%
  pivot_longer(
    cols = all_of(genes_to_plot),
    names_to = "gene",
    values_to = "expression"
  ) %>%
  # Filter for cells that express the gene (expression > 0)
  filter(expression > 0)

# Randomly shuffle the rows of the combined data frame
shuffled_data <- long_data %>%
  sample_frac(1)

ggplot() +
  # 1. Plot all cells as a light grey background layer
  geom_point(data = background_data,
             aes(x = umapharmony_1, y = umapharmony_2),
             color = "lightgrey",
             size = 0.5,
             alpha = 0.5) +
  
  # 2. Plot the shuffled, expressing cells on top
  geom_point(data = shuffled_data,
             aes(x = umapharmony_1, y = umapharmony_2, color = gene),
             alpha = 0.5,
             size = 0.8) +
  
  # Set manual colors and customize the theme, also, remember to replace genes with those you are interested in
  scale_color_manual(values = c(
    "SLC17A6" = "orange",
    "SLC17A7" = "blue")) +
  # Use guides() to override the size of the points in the legend
  guides(color = guide_legend(override.aes = list(size = 4))) +
  labs(title = "SLC17A6, SLC17A7",
       color = "Gene Expression") +
  theme_minimal() +
  theme(
    panel.grid = element_blank(), # Remove grid lines
    axis.line = element_line(color = "black"), # Add axis lines back in
    # Increase legend title and text size
    legend.title = element_text(size = 14), 
    legend.text = element_text(size = 12)
  )

#### Astro work ####

# Generate the UMAP plot
plot <- DimPlot(object = astro_sub, reduction = "umap.harmony") # replace neuro_sub with subset of interest

# Pass the plot to CellSelector()
# The R console will prompt you to select points on the plot
cells.located <- CellSelector(plot = plot)
# Create a new identity class for the selected cells
astro_sub <- SetIdent(astro_sub, cells = cells.located, value = 'selected_cells')


astro_sub <- RenameIdents(
  object = astro_sub,
  'selected_cells' = "CombinedGroup"
)
# Verify the selection with a new plot
DimPlot(astro_sub, reduction = "umap.harmony", group.by = 'ident')

astro_sub$lasso <- Idents(astro_sub)

# Correctly create the new metadata column using as.character()
astro_sub$cell_type_updated <- ifelse(
  test = astro_sub$lasso == "CombinedGroup",
  yes = as.character(astro_sub$cell_type_v25),
  no = "other"
)

# You can convert the column back to a factor if needed for plotting
# R will automatically convert character vectors to factors when they are used for grouping
# so this is often not required for standard plotting functions
astro_sub$cell_type_updated <- as.factor(astro_sub$cell_type_updated)

# Check the new assignments
table(astro_sub$cell_type_updated)

# Now, plot with the new, correctly labeled column
DimPlot_scCustom(astro_sub, group.by = "cell_type_updated", reduction = "umap.harmony")

Idents(astro_sub) <- "cell_type_updated"
subset_to_plot <- subset(astro_sub, idents = "other", invert = TRUE)

DimPlot_scCustom(subset_to_plot, group.by = "cell_type_updated", reduction = "umap.harmony")
DimPlot_scCustom(subset_to_plot, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = TRUE), color_seed = 5)

saveRDS(astro_sub, 'astro_sub.RDS')

### Plot genes of interest ###
Idents(astro_sub) <- "cell_type_updated"
subset_to_plot <- subset(astro_sub, idents = "other", invert = TRUE) # replace neur_sub with subset of interest
umap_coords <- as.data.frame(subset_to_plot@reductions$umap.harmony@cell.embeddings)

# Ensure the genes exist in your object before proceeding
genes_to_plot <- c("SLC17A6", "SLC17A7") # replace with your genes of interest
if (!all(genes_to_plot %in% rownames(fb_seurat))) {
  stop("One or more genes not found in the Seurat object.")
}

# Fetch expression data. `FetchData` is the safest way to get expression.
expression_data <- FetchData(object = subset_to_plot, vars = genes_to_plot)

# Data for the background layer (all cells)
background_data <- umap_coords

# Prepare data for the shuffled foreground layers (expressing cells only)
plot_data <- cbind(umap_coords, expression_data)

long_data <- plot_data %>%
  pivot_longer(
    cols = all_of(genes_to_plot),
    names_to = "gene",
    values_to = "expression"
  ) %>%
  # Filter for cells that express the gene (expression > 0)
  filter(expression > 0)

# Randomly shuffle the rows of the combined data frame
shuffled_data <- long_data %>%
  sample_frac(1)

ggplot() +
  # 1. Plot all cells as a light grey background layer
  geom_point(data = background_data,
             aes(x = umapharmony_1, y = umapharmony_2),
             color = "lightgrey",
             size = 0.5,
             alpha = 0.5) +
  
  # 2. Plot the shuffled, expressing cells on top
  geom_point(data = shuffled_data,
             aes(x = umapharmony_1, y = umapharmony_2, color = gene),
             alpha = 0.5,
             size = 0.8) +
  
  # Set manual colors and customize the theme, also, remember to replace genes with those you are interested in
  scale_color_manual(values = c(
    "SLC17A6" = "orange",
    "SLC17A7" = "blue")) +
  # Use guides() to override the size of the points in the legend
  guides(color = guide_legend(override.aes = list(size = 4))) +
  labs(title = "SLC17A6, SLC17A7",
       color = "Gene Expression") +
  theme_minimal() +
  theme(
    panel.grid = element_blank(), # Remove grid lines
    axis.line = element_line(color = "black"), # Add axis lines back in
    # Increase legend title and text size
    legend.title = element_text(size = 14), 
    legend.text = element_text(size = 12)
  )


#### Micro work ####
levels(fb_seurat)
clustersToSubset <- c("MG")
micro_sub <- subset(fb_seurat, idents = clustersToSubset)
saveRDS(micro_sub, "micro_sub.RDS")


# Generate the UMAP plot
plot <- DimPlot(object = micro_sub, reduction = "umap.harmony") # replace neuro_sub with subset of interest

# Pass the plot to CellSelector()
# The R console will prompt you to select points on the plot
cells.located <- CellSelector(plot = plot)
# Create a new identity class for the selected cells
micro_sub <- SetIdent(micro_sub, cells = cells.located, value = 'selected_cells')

micro_sub <- RenameIdents(
  object = micro_sub,
  'selected_cells' = "CombinedGroup"
)
# Verify the selection with a new plot
DimPlot(micro_sub, reduction = "umap.harmony", group.by = 'ident')

micro_sub$lasso <- Idents(micro_sub)

# Correctly create the new metadata column using as.character()
micro_sub$cell_type_updated <- ifelse(
  test = micro_sub$lasso == "CombinedGroup",
  yes = as.character(micro_sub$cell_type_v25),
  no = "other"
)

# You can convert the column back to a factor if needed for plotting
# R will automatically convert character vectors to factors when they are used for grouping
# so this is often not required for standard plotting functions
micro_sub$cell_type_updated <- as.factor(micro_sub$cell_type_updated)

# Check the new assignments
table(micro_sub$cell_type_updated)

# Now, plot with the new, correctly labeled column
DimPlot_scCustom(micro_sub, group.by = "cell_type_updated", reduction = "umap.harmony")

Idents(micro_sub) <- "cell_type_updated"
subset_to_plot <- subset(micro_sub, idents = "other", invert = TRUE)

DimPlot_scCustom(subset_to_plot, group.by = "cell_type_updated", reduction = "umap.harmony")
DimPlot_scCustom(subset_to_plot, reduction = "umap.harmony", label = TRUE, label.box = TRUE, pt.size = 0.5, repel = TRUE, shuffle = TRUE, colors_use = DiscretePalette_scCustomize(num_colors = 29, palette = "varibow", shuffle_pal = TRUE), color_seed = 5)

saveRDS(micro_sub, 'micro_sub.RDS')

### Plot genes of interest ###
Idents(micro_sub) <- "cell_type_updated"
subset_to_plot <- subset(micro_sub, idents = "other", invert = TRUE) # replace neur_sub with subset of interest
umap_coords <- as.data.frame(subset_to_plot@reductions$umap.harmony@cell.embeddings)

# Ensure the genes exist in your object before proceeding
genes_to_plot <- c("SLC17A6", "SLC17A7") # replace with your genes of interest
if (!all(genes_to_plot %in% rownames(fb_seurat))) {
  stop("One or more genes not found in the Seurat object.")
}

# Fetch expression data. `FetchData` is the safest way to get expression.
expression_data <- FetchData(object = subset_to_plot, vars = genes_to_plot)

# Data for the background layer (all cells)
background_data <- umap_coords

# Prepare data for the shuffled foreground layers (expressing cells only)
plot_data <- cbind(umap_coords, expression_data)

long_data <- plot_data %>%
  pivot_longer(
    cols = all_of(genes_to_plot),
    names_to = "gene",
    values_to = "expression"
  ) %>%
  # Filter for cells that express the gene (expression > 0)
  filter(expression > 0)

# Randomly shuffle the rows of the combined data frame
shuffled_data <- long_data %>%
  sample_frac(1)

ggplot() +
  # 1. Plot all cells as a light grey background layer
  geom_point(data = background_data,
             aes(x = umapharmony_1, y = umapharmony_2),
             color = "lightgrey",
             size = 0.5,
             alpha = 0.5) +
  
  # 2. Plot the shuffled, expressing cells on top
  geom_point(data = shuffled_data,
             aes(x = umapharmony_1, y = umapharmony_2, color = gene),
             alpha = 0.5,
             size = 0.8) +
  
  # Set manual colors and customize the theme, also, remember to replace genes with those you are interested in
  scale_color_manual(values = c(
    "SLC17A6" = "orange",
    "SLC17A7" = "blue")) +
  # Use guides() to override the size of the points in the legend
  guides(color = guide_legend(override.aes = list(size = 4))) +
  labs(title = "SLC17A6, SLC17A7",
       color = "Gene Expression") +
  theme_minimal() +
  theme(
    panel.grid = element_blank(), # Remove grid lines
    axis.line = element_line(color = "black"), # Add axis lines back in
    # Increase legend title and text size
    legend.title = element_text(size = 14), 
    legend.text = element_text(size = 12)
  )


#### Adding Gestational age at delivery ####
head(fb_seurat@meta.data)
unique(fb_seurat$sample)
fb_seurat$gest_age <- fb_seurat$sample
Idents(fb_seurat) <- "gest_age"
Idents(fb_seurat)

fb_seurat <- RenameIdents(fb_seurat, c("Ctrl13" = "158.4", "Ctrl24" = "147.4", "Ctrl25" = "144.3", "Ctrl26" = "157.4", 'Ctrl27' = "148.4", "Sal7" = "135.4", "Sal9" = "128.4", "Sal10" = "134.4", "Sal13" = "132.3"))
fb_seurat$gest_age <- Idents(fb_seurat)

# Get cell counts per cluster per sample
cell_counts <- table(fb_seurat@meta.data$cell_type_v25, fb_seurat@meta.data$gest_age)
cell_counts <- as.data.frame(cell_counts)
cell_counts <- cell_counts[order(cell_counts$Var2, decreasing = TRUE), ]

cell_counts_df <- as.data.frame(cell_counts)
cell_counts_sorted <- cell_counts_df %>%
  mutate(
    # Reorder the factor levels of Var2 based on their numeric value
    Var2 = fct_relevel(Var2, sort(as.character(levels(Var2)), decreasing = FALSE))
  ) %>%
  # Arrange the data frame by the new level order for proper display
  arrange(Var2)

# Check the new level order
levels(cell_counts_sorted$Var2)
# View the sorted data frame
head(cell_counts_sorted)


ggplot(data=cell_counts_sorted, aes(x=Var2, y=Freq, fill = Var1)) +
  geom_bar(stat="identity", position="fill") + 
  labs(title = "Cell Proportion Per Sample, Ranked by Gestational Age", x= "Gestational Age", y= "Cell Proportion") + 
  guides(fill = guide_legend(title = "Cell Type",reverse=FALSE)) +
  #scale_fill_discrete(labels = c("CV", "Decidua", "Meminoc")) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggplot(data = cell_counts_sorted, aes(x = Var2, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "fill", color = "black") + 
  labs(title = "Cell Proportion Per Sample, Ranked by Gestational Age", x = "Gestational Age", y = "Cell Proportion") + 
  guides(fill = guide_legend(title = "Cell Type", reverse = FALSE)) +
  #scale_fill_discrete(labels = c("CV", "Decidua", "Meminoc")) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggplot(data = cell_counts_sorted, aes(x = Var2, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "fill", color = "black") + 
  labs(title = "Cell Proportion Per Sample, Ranked by Gestational Age", x = "Gestational Age", y = "Cell Proportion") + 
  guides(fill = guide_legend(title = "Cell Type", reverse = FALSE)) +
  # Use a ColorBrewer palette, for example "Set1"
  scale_fill_brewer(palette = "Dark2") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggplot(data = cell_counts_sorted, aes(x = Var2, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "fill", color = "black") + 
  labs(title = "Cell Proportion Per Sample, Ranked by Gestational Age", x = "Gestational Age", y = "Cell Proportion") + 
  guides(fill = guide_legend(title = "Cell Type", reverse = FALSE)) +
  # Use a viridis palette, for example "plasma"
  scale_fill_viridis_d(option = "plasma") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggplot(data = cell_counts_sorted, aes(x = Var2, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "fill", color = "black") + 
  labs(title = "Cell Proportion Per Sample, Ranked by Gestational Age", x = "Gestational Age", y = "Cell Proportion") + 
  guides(fill = guide_legend(title = "Cell Type", reverse = FALSE)) +
  scale_fill_paletteer_d("khroma::bright") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Get a starting palette (e.g., Set1) and create a function to interpolate more colors
get_28_colors <- colorRampPalette(brewer.pal(9, "Set1"))

ggplot(data = cell_counts_sorted, aes(x = Var2, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "fill", color = "black") + 
  labs(title = "Cell Proportion Per Sample, Ranked by Gestational Age", x = "Gestational Age", y = "Cell Proportion") + 
  guides(fill = guide_legend(title = "Cell Type", reverse = FALSE)) +
  scale_fill_manual(values = get_28_colors(28)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

head(cell_counts_sorted)

# First, modify and aggregate the data to create the new grouped categories
cell_counts_binned <- cell_counts_sorted %>%
  mutate(Var1_grouped = case_when(
    startsWith(as.character(Var1), "ExN") ~ "ExN",
    Var1 %in% c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1") ~ "G.E. Neuro",
    Var1 %in% c("OPC0", "OPC1", "OPC2", "COP", "OL1", "OL2", "AS0", "AS1", "AS2") ~ "Oligo/Astro",
    TRUE ~ as.character(Var1)
  )) %>%
  # Aggregate the data by the new grouped variable and the x-axis variable
  group_by(Var2, Var1_grouped) %>%
  summarise(Freq = sum(Freq), .groups = "drop")

# Now, plot the aggregated data
ggplot(data = cell_counts_binned, aes(x = Var2, y = Freq, fill = Var1_grouped)) +
  geom_bar(stat = "identity", position = "fill", color = "black") + 
  labs(title = "Cell Proportion Per Sample, Ranked by Gestational Age", x = "Gestational Age", y = "Cell Proportion") + 
  guides(fill = guide_legend(title = "Cell Type", reverse = FALSE)) +
  scale_fill_brewer(palette = "Set3") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


# First, modify and aggregate the data to create the new grouped categories
cell_counts_binned <- cell_counts_sorted %>%
  mutate(Var1_grouped = case_when(
    startsWith(as.character(Var1), "ExN") ~ "ExN",
    Var1 %in% c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1") ~ "G.E. Neuro",
    Var1 %in% c("OPC0", "OPC1", "OPC2", "COP", "OL1", "OL2", "AS0", "AS1", "AS2") ~ "Oligo/Astro",
    TRUE ~ as.character(Var1)
  )) %>%
  # Aggregate the data by the new grouped variable and the x-axis variable
  group_by(Var2, Var1_grouped) %>%
  summarise(Freq = sum(Freq), .groups = "drop")

# Now, plot the aggregated data and use facet_wrap to separate the plots
ggplot(data = cell_counts_binned, aes(x = Var2, y = Freq, fill = Var1_grouped)) +
  geom_bar(stat = "identity", position = "fill", color = "black") + 
  labs(title = "Cell Proportion Per Sample, Ranked by Gestational Age", x = "Gestational Age", y = "Cell Proportion") + 
  guides(fill = guide_legend(title = "Cell Type", reverse = FALSE)) +
  scale_fill_brewer(palette = "Set3") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  # This is the new line to create separate plots for each group
  facet_wrap(~Var1_grouped)


# First, modify and aggregate the data to create the new grouped categories
# You do not need to aggregate the data this time, as you want to keep the original labels separate.
cell_counts_binned <- cell_counts_sorted %>%
  mutate(Var1_grouped = case_when(
    startsWith(as.character(Var1), "ExN") ~ "ExN",
    Var1 %in% c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1") ~ "G.E. Neuro",
    Var1 %in% c("OPC0", "OPC1", "OPC2", "COP", "OL1", "OL2", "AS0", "AS1", "AS2") ~ "Oligo/Astro",
    TRUE ~ as.character(Var1)
  ))

# Now, plot the unaggregated data, using the original labels for fill and the new group for facetting
ggplot(data = cell_counts_binned, aes(x = Var2, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "fill", color = "black") + 
  labs(title = "Cell Proportion Per Sample, Ranked by Gestational Age", x = "Gestational Age", y = "Cell Proportion") + 
  guides(fill = guide_legend(title = "Cell Type", reverse = FALSE)) +
  # Use a color palette that supports the large number of original categories
  scale_fill_brewer(palette = "Paired") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  # Keep the facet_wrap to split by the grouped categories
  facet_wrap(~Var1_grouped)


# Define the list of cell types to remove
cell_types_to_remove <- c("EP", "Mono", "MG", "PC", "EC", "EB0", "EB1", "CD4+T")

# First, modify the data and filter out the unwanted cell types
cell_counts_binned <- cell_counts_sorted %>%
  filter(!Var1 %in% cell_types_to_remove) %>%
  mutate(Var1_grouped = case_when(
    startsWith(as.character(Var1), "ExN") ~ "ExN",
    Var1 %in% c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1") ~ "G.E. Neuro",
    Var1 %in% c("OPC0", "OPC1", "OPC2", "COP", "OL1", "OL2", "AS0", "AS1", "AS2") ~ "Oligo/Astro",
    TRUE ~ as.character(Var1)
  ))

# Now, plot the filtered data
ggplot(data = cell_counts_binned, aes(x = Var2, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "fill", color = "black") + 
  labs(title = "Cell Proportion Per Sample, Ranked by Gestational Age", x = "Gestational Age", y = "Cell Proportion") + 
  guides(fill = guide_legend(title = "Cell Type", reverse = FALSE)) +
  scale_fill_brewer(palette = "Paired") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  facet_wrap(~Var1_grouped)





# Define the list of cell types to remove
cell_types_to_remove <- c("EP", "Mono", "MG", "PC", "EC", "EB0", "EB1", "CD4+T")

# Filter the data and create the new grouped variable for facetting
cell_counts_binned <- cell_counts_sorted %>%
  filter(!Var1 %in% cell_types_to_remove) %>%
  mutate(Var1_grouped = case_when(
    startsWith(as.character(Var1), "ExN") ~ "ExN",
    Var1 %in% c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1") ~ "G.E. Neuro",
    Var1 %in% c("OPC0", "OPC1", "OPC2", "COP", "OL1", "OL2", "AS0", "AS1", "AS2") ~ "Oligo/Astro",
    TRUE ~ as.character(Var1)
  ))

# Use the `khroma` package for a large, colorblind-safe palette
# `khroma::kelly()` provides up to 22 colors that are easy to distinguish
my_palette <- khroma::color("discreterainbow")(length(unique(cell_counts_binned$Var1)))

# Create the plot using the new palette
ggplot(data = cell_counts_binned, aes(x = Var2, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "fill", color = "black") + 
  labs(title = "Cell Proportion Per Sample, Ranked by Gestational Age", x = "Gestational Age", y = "Cell Proportion") + 
  guides(fill = guide_legend(title = "Cell Type", reverse = FALSE)) +
  scale_fill_manual(values = my_palette) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  facet_wrap(~Var1_grouped)



# Define the list of cell types to remove
cell_types_to_remove <- c("EP", "Mono", "MG", "PC", "EC", "EB0", "EB1", "CD4+T")

# Filter the data and create the new grouped variable for facetting
cell_counts_binned <- cell_counts_sorted %>%
  filter(!Var1 %in% cell_types_to_remove) %>%
  mutate(Var1_grouped = case_when(
    Var1 %in% c("AS0", "AS1", "AS2") ~ "Astro",
    # Renaming "Oligo/Astro" to "Oligo"
    Var1 %in% c("OPC0", "OPC1", "OPC2", "COP", "OL1", "OL2") ~ "Oligo",
    startsWith(as.character(Var1), "ExN") ~ "ExN",
    Var1 %in% c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1") ~ "G.E. Neuro",
    TRUE ~ as.character(Var1)
  ))

# Create a function to generate a palette of any size
get_colors <- colorRampPalette(brewer.pal(9, "Set1"))

# Determine the number of colors needed
num_colors <- length(unique(cell_counts_binned$Var1))
my_palette <- get_colors(num_colors)

# Create the plot using the new palette
ggplot(data = cell_counts_binned, aes(x = Var2, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "fill", color = "black") + 
  labs(title = "Cell Proportion Per Sample, Ranked by Gestational Age", x = "Gestational Age", y = "Cell Proportion") + 
  guides(fill = guide_legend(title = "Cell Type", reverse = FALSE)) +
  scale_fill_manual(values = my_palette) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  facet_wrap(~Var1_grouped)




# Define the list of cell types to remove
cell_types_to_remove <- c("EP", "Mono", "MG", "PC", "EC", "EB0", "EB1", "CD4+T")

# Filter the data and create the new grouped variable for facetting
cell_counts_binned <- cell_counts_sorted %>%
  filter(!Var1 %in% cell_types_to_remove) %>%
  mutate(Var1_grouped = case_when(
    Var1 %in% c("AS0", "AS1", "AS2") ~ "Astro",
    # Renaming "Oligo/Astro" to "Oligo"
    Var1 %in% c("OPC0", "OPC1", "OPC2", "COP", "OL1", "OL2") ~ "Oligo",
    startsWith(as.character(Var1), "ExN") ~ "ExN",
    Var1 %in% c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1") ~ "G.E. Neuro",
    TRUE ~ as.character(Var1)
  ))

# Create a function to generate a palette of any size
get_colors <- colorRampPalette(brewer.pal(9, "Set1"))

# Get all unique cell types from the filtered data
all_cell_types <- unique(cell_counts_binned$Var1)

# Get default colors for all types
default_colors <- get_colors(length(all_cell_types))
names(default_colors) <- all_cell_types

# Define custom colors for the specific types
my_custom_colors <- default_colors
my_custom_colors["COP"] <- "red"
my_custom_colors["OL1"] <- "darkorange"
my_custom_colors["vRG"] <- "dodgerblue"
my_custom_colors["oRG"] <- "skyblue"

# Create the plot using the custom palette
ggplot(data = cell_counts_binned, aes(x = Var2, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "fill", color = "black") +
  labs(title = "Cell Proportion Per Sample, Ranked by Gestational Age", x = "Gestational Age", y = "Cell Proportion") +
  guides(fill = guide_legend(title = "Cell Type", reverse = FALSE)) +
  scale_fill_manual(values = my_custom_colors) + # Use the new custom palette
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  facet_wrap(~Var1_grouped)





# Define the list of cell types to remove
cell_types_to_remove <- c("EP", "Mono", "MG", "PC", "EC", "EB0", "EB1", "CD4+T")

# Filter the data and create the new grouped variable for facetting
cell_counts_binned <- cell_counts_sorted %>%
  filter(!Var1 %in% cell_types_to_remove) %>%
  mutate(Var1_grouped = case_when(
    Var1 %in% c("AS0", "AS1", "AS2") ~ "Astro",
    # Renaming "Oligo/Astro" to "Oligo"
    Var1 %in% c("OPC0", "OPC1", "OPC2", "COP", "OL1", "OL2") ~ "Oligo",
    startsWith(as.character(Var1), "ExN") ~ "ExN",
    Var1 %in% c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1") ~ "G.E. Neuro",
    TRUE ~ as.character(Var1)
  ))

# Create a function to generate a palette of any size
get_colors <- colorRampPalette(brewer.pal(9, "Set1"))

# Get all unique cell types from the filtered data
all_cell_types <- unique(cell_counts_binned$Var1)

# Get default colors for all types
default_colors <- get_colors(length(all_cell_types))
names(default_colors) <- all_cell_types

# Define custom colors for the specific types
my_custom_colors <- default_colors
my_custom_colors["COP"] <- "red"
my_custom_colors["OL1"] <- "darkorange"
my_custom_colors["vRG"] <- "dodgerblue"
my_custom_colors["oRG"] <- "skyblue"

# Create the plot using the custom palette and bold text theme
ggplot(data = cell_counts_binned, aes(x = Var2, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "fill", color = "black") +
  labs(title = "Cell Proportion Per Sample, Ranked by Gestational Age", x = "Gestational Age", y = "Cell Proportion") +
  guides(fill = guide_legend(title = "Cell Type", reverse = FALSE)) +
  scale_fill_manual(values = my_custom_colors) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    # Make all text elements bold
    text = element_text(face = "bold")
  ) +
  facet_wrap(~Var1_grouped)







# Define the list of cell types to remove
cell_types_to_remove <- c("EP", "Mono", "MG", "PC", "EC", "EB0", "EB1", "CD4+T")

# Filter the data and create the new grouped variable for facetting
cell_counts_binned <- cell_counts_sorted %>%
  filter(!Var1 %in% cell_types_to_remove) %>%
  mutate(Var1_grouped = case_when(
    Var1 %in% c("AS0", "AS1", "AS2") ~ "Astro",
    # Renaming "Oligo/Astro" to "Oligo"
    Var1 %in% c("OPC0", "OPC1", "OPC2", "COP", "OL1", "OL2") ~ "Oligo",
    startsWith(as.character(Var1), "ExN") ~ "ExN",
    Var1 %in% c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1") ~ "G.E. Neuro",
    TRUE ~ as.character(Var1)
  ))

# Create a function to generate a palette of any size
get_colors <- colorRampPalette(brewer.pal(9, "Set1"))

# Get all unique cell types from the filtered data
all_cell_types <- unique(cell_counts_binned$Var1)

# Get default colors for all types
default_colors <- get_colors(length(all_cell_types))
names(default_colors) <- all_cell_types

# Define custom colors for the specific types
my_custom_colors <- default_colors
my_custom_colors["COP"] <- "red"
my_custom_colors["OL1"] <- "darkorange"
my_custom_colors["vRG"] <- "dodgerblue"
my_custom_colors["oRG"] <- "skyblue"

# Create the plot with the custom palette, bold text, and a two-column legend
ggplot(data = cell_counts_binned, aes(x = Var2, y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", position = "fill", color = "black") +
  labs(title = "Cell Proportion Per Sample, Ranked by Gestational Age", x = "Gestational Age", y = "Cell Proportion") +
  guides(fill = guide_legend(title = "Cell Type", reverse = FALSE, ncol = 2)) + # Set legend to two columns
  scale_fill_manual(values = my_custom_colors) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    text = element_text(face = "bold")
  ) +
  facet_wrap(~Var1_grouped)














# Define the list of cell types to remove
cell_types_to_remove <- c("EP", "Mono", "MG", "PC", "EC", "EB0", "EB1", "CD4+T")

# Filter the data and create the new grouped variable for facetting
cell_counts_binned <- cell_counts_sorted %>%
  filter(!Var1 %in% cell_types_to_remove) %>%
  mutate(Var1_grouped = case_when(
    Var1 %in% c("AS0", "AS1", "AS2") ~ "Astro",
    Var1 %in% c("OPC0", "OPC1", "OPC2", "COP", "OL1", "OL2") ~ "Oligo",
    startsWith(as.character(Var1), "ExN") ~ "ExN",
    Var1 %in% c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1") ~ "G.E. Neuro",
    TRUE ~ as.character(Var1)
  ))

# Define custom colors
my_custom_colors <- c(
  "COP" = "red",
  "OL1" = "darkorange",
  "vRG" = "dodgerblue",
  "oRG" = "skyblue"
)

# Get all unique cell types
all_cell_types <- unique(cell_counts_binned$Var1)
default_colors <- colorRampPalette(brewer.pal(9, "Set1"))(length(all_cell_types))
names(default_colors) <- all_cell_types

# Combine custom colors with default colors
final_colors <- default_colors
final_colors[names(my_custom_colors)] <- my_custom_colors

# Create a list to hold the individual plots
plot_list <- list()

# Get the order of facets
facet_order <- unique(cell_counts_binned$Var1_grouped)

# Generate a plot for each facet group
for (group_name in facet_order) {
  # Filter data for the current group
  group_data <- cell_counts_binned %>% filter(Var1_grouped == group_name)
  
  # Get the colors relevant to this facet only
  group_colors <- final_colors[names(final_colors) %in% unique(group_data$Var1)]
  
  # Create the plot for this facet
  p <- ggplot(data = group_data, aes(x = Var2, y = Freq, fill = Var1)) +
    geom_bar(stat = "identity", position = "fill", color = "black") +
    labs(title = group_name, x = "", y = "") + # Use group_name as title
    guides(fill = guide_legend(title = "Cell Type", ncol = 1)) +
    scale_fill_manual(values = group_colors) +
    theme(
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      text = element_text(face = "bold"),
      legend.position = "right",
      plot.title = element_text(hjust = 0.5)
    )
  
  plot_list[[group_name]] <- p
}

# Combine the individual plots using patchwork
# We need a shared title and axis labels for the combined plot
combined_plot <- wrap_plots(plot_list, ncol = 2) +
  plot_annotation(
    title = "Cell Proportion Per Sample, Ranked by Gestational Age",
    tag_levels = 'A',
    theme = theme(text = element_text(face = "bold"))
  ) &
  theme(plot.title = element_text(hjust = 0.5))

# Display the combined plot
combined_plot






# Define the list of cell types to remove
cell_types_to_remove <- c("EP", "Mono", "MG", "PC", "EC", "EB0", "EB1", "CD4+T")

# Filter the data and create the new grouped variable for facetting
cell_counts_binned <- cell_counts_sorted %>%
  filter(!Var1 %in% cell_types_to_remove) %>%
  mutate(Var1_grouped = case_when(
    Var1 %in% c("AS0", "AS1", "AS2") ~ "Astro",
    Var1 %in% c("OPC0", "OPC1", "OPC2", "COP", "OL1", "OL2") ~ "Oligo",
    startsWith(as.character(Var1), "ExN") ~ "ExN",
    Var1 %in% c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1") ~ "G.E. Neuro",
    TRUE ~ as.character(Var1)
  ))

# Define custom colors
my_custom_colors <- c(
  "COP" = "red",
  "OL1" = "darkorange",
  "vRG" = "dodgerblue",
  "oRG" = "skyblue"
)

# Get all unique cell types
all_cell_types <- unique(cell_counts_binned$Var1)
default_colors <- colorRampPalette(brewer.pal(9, "Set1"))(length(all_cell_types))
names(default_colors) <- all_cell_types

# Combine custom colors with default colors
final_colors <- default_colors
final_colors[names(my_custom_colors)] <- my_custom_colors

# Create a list to hold the individual plots
plot_list <- list()
facet_order <- unique(cell_counts_binned$Var1_grouped)
num_facets <- length(facet_order)

# Set up the layout
# Assuming a 2x2 layout, the plots in the top row should have their x-axis labels removed.
# The plots in the bottom row should keep their labels.
layout_ncol <- 2 # Adjust as needed
plots_in_row <- layout_ncol
num_rows <- ceiling(num_facets / plots_in_row)
plots_with_labels <- ((num_rows - 1) * plots_in_row + 1):num_facets

# Generate a plot for each facet group
for (i in 1:num_facets) {
  group_name <- facet_order[i]
  group_data <- cell_counts_binned %>% filter(Var1_grouped == group_name)
  group_colors <- final_colors[names(final_colors) %in% unique(group_data$Var1)]
  
  p <- ggplot(data = group_data, aes(x = Var2, y = Freq, fill = Var1)) +
    geom_bar(stat = "identity", position = "fill", color = "black") +
    labs(title = group_name, x = "", y = "") +
    guides(fill = guide_legend(title = "Cell Type", ncol = 1)) +
    scale_fill_manual(values = group_colors) +
    theme(
      text = element_text(face = "bold"),
      legend.position = "right",
      plot.title = element_text(hjust = 0.5)
    )
  
  # Keep x-axis labels only for the bottom plots
  if (!(i %in% plots_with_labels)) {
    p <- p + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())
  } else {
    p <- p + theme(axis.text.x = element_text(angle = 45, hjust = 1))
  }
  
  plot_list[[group_name]] <- p
}

# Add shared title and axis labels using patchwork
combined_plot <- wrap_plots(plot_list, ncol = layout_ncol) +
  plot_annotation(
    title = "Cell Proportion Per Sample, Ranked by Gestational Age",
    theme = theme(text = element_text(face = "bold"), plot.title = element_text(hjust = 0.5))
  ) &
  plot_layout(guides = "collect") &
  theme(plot.title = element_text(hjust = 0.5))

# Add a common x-axis label at the bottom using plot_annotation
combined_plot <- combined_plot + plot_annotation(
  caption = "Gestational Age",
  theme = theme(
    plot.caption.position = "plot",
    plot.caption = element_text(face = "bold", hjust = 0.5, size = 12)
  )
)

# Display the combined plot
combined_plot






# Define the list of cell types to remove
cell_types_to_remove <- c("EP", "Mono", "MG", "PC", "EC", "EB0", "EB1", "CD4+T")

# Filter the data and create the new grouped variable for facetting
cell_counts_binned <- cell_counts_sorted %>%
  filter(!Var1 %in% cell_types_to_remove) %>%
  mutate(Var1_grouped = case_when(
    Var1 %in% c("AS0", "AS1", "AS2") ~ "Astro",
    Var1 %in% c("OPC0", "OPC1", "OPC2", "COP", "OL1", "OL2") ~ "Oligo",
    startsWith(as.character(Var1), "ExN") ~ "ExN",
    Var1 %in% c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1") ~ "G.E. Neuro",
    TRUE ~ as.character(Var1)
  ))

# Define custom colors
my_custom_colors <- c(
  "COP" = "red",
  "OL1" = "darkorange",
  "vRG" = "dodgerblue",
  "oRG" = "skyblue"
)

# Get all unique cell types
all_cell_types <- unique(cell_counts_binned$Var1)
default_colors <- colorRampPalette(brewer.pal(9, "Set1"))(length(all_cell_types))
names(default_colors) <- all_cell_types

# Combine custom colors with default colors
final_colors <- default_colors
final_colors[names(my_custom_colors)] <- my_custom_colors

# Create a list to hold the individual plots
plot_list <- list()
facet_order <- unique(cell_counts_binned$Var1_grouped)
num_facets <- length(facet_order)

# Generate a plot for each facet group
for (i in 1:num_facets) {
  group_name <- facet_order[i]
  group_data <- cell_counts_binned %>% filter(Var1_grouped == group_name)
  group_colors <- final_colors[names(final_colors) %in% unique(group_data$Var1)]
  
  p <- ggplot(data = group_data, aes(x = Var2, y = Freq, fill = Var1)) +
    geom_bar(stat = "identity", position = "fill", color = "black") +
    labs(title = group_name, x = "Gestational Age", y = "Cell Proportion") +
    guides(fill = guide_legend(title = "Cell Type", ncol = 1)) +
    scale_fill_manual(values = group_colors) +
    theme(
      text = element_text(face = "bold"),
      legend.position = "right",
      plot.title = element_text(hjust = 0.5),
      axis.text.x = element_text(angle = 45, hjust = 1, size = 8), # Include x-axis text
      axis.ticks.x = element_line() # Include x-axis ticks
    )
  
  plot_list[[group_name]] <- p
}

# Combine the individual plots using patchwork
combined_plot <- wrap_plots(plot_list, ncol = 2) +
  plot_annotation(
    title = "Cell Proportion Per Sample, Ranked by Gestational Age",
    theme = theme(text = element_text(face = "bold"), plot.title = element_text(hjust = 0.5))
  ) &
  plot_layout(guides = "collect") &
  theme(plot.title = element_text(hjust = 0.5))

# Display the combined plot
combined_plot








# Define the list of cell types to remove
cell_types_to_remove <- c("EP", "Mono", "MG", "PC", "EC", "EB0", "EB1", "CD4+T")

# Filter the data and create the new grouped variable for facetting
cell_counts_binned <- cell_counts_sorted %>%
  filter(!Var1 %in% cell_types_to_remove) %>%
  mutate(Var1_grouped = case_when(
    Var1 %in% c("AS0", "AS1", "AS2") ~ "Astro",
    Var1 %in% c("OPC0", "OPC1", "OPC2", "COP", "OL1", "OL2") ~ "Oligo",
    startsWith(as.character(Var1), "ExN") ~ "ExN",
    Var1 %in% c("vRG", "oRG", "CGE0", "CGE1", "CGE2", "MGE0", "MGE1") ~ "G.E. Neuro",
    TRUE ~ as.character(Var1)
  ))

# Define custom colors
my_custom_colors <- c(
  "COP" = "red",
  "OL1" = "darkorange",
  "vRG" = "dodgerblue",
  "oRG" = "skyblue"
)

# Get all unique cell types
all_cell_types <- unique(cell_counts_binned$Var1)
default_colors <- colorRampPalette(brewer.pal(9, "Set1"))(length(all_cell_types))
names(default_colors) <- all_cell_types

# Combine custom colors with default colors
final_colors <- default_colors
final_colors[names(my_custom_colors)] <- my_custom_colors

# Create a list to hold the individual plots
plot_list <- list()
facet_order <- unique(cell_counts_binned$Var1_grouped)
num_facets <- length(facet_order)

# Generate a plot for each facet group
for (i in 1:num_facets) {
  group_name <- facet_order[i]
  group_data <- cell_counts_binned %>% filter(Var1_grouped == group_name)
  group_colors <- final_colors[names(final_colors) %in% unique(group_data$Var1)]
  
  p <- ggplot(data = group_data, aes(x = Var2, y = Freq, fill = Var1)) +
    geom_bar(stat = "identity", position = "fill", color = "black") +
    labs(title = group_name, x = "Gestational Age", y = "Cell Proportion") +
    guides(fill = guide_legend(title = "Cell Type", ncol = 1)) +
    scale_fill_manual(values = group_colors) +
    theme(
      text = element_text(face = "bold"),
      legend.position = "right", # Keep legend next to its plot
      plot.title = element_text(hjust = 0.5),
      axis.text.x = element_text(angle = 45, hjust = 1, size = 8),
      axis.ticks.x = element_line()
    )
  
  plot_list[[group_name]] <- p
}

# Combine the plots without collecting legends
# This will arrange the plots in a 2x2 grid and each will have its own legend
# You can use plot_list[["ExN"]] + plot_list[["G.E. Neuro"]] + ... for manual arrangement
# A cleaner way is to use a layout and wrap_plots
layout <- "
  AABB
  CCDD
"
combined_plot <- wrap_plots(plot_list, design = layout)

# Add a title and theme for the overall figure
combined_plot <- combined_plot + plot_annotation(
  title = "Cell Proportion Per Sample, Ranked by Gestational Age",
  theme = theme(
    text = element_text(face = "bold"),
    plot.title = element_text(hjust = 0.5, size = 16)
  )
)

# Display the combined plot
combined_plot